{"cells":[{"cell_type":"markdown","metadata":{"id":"OTA2scaimmBA"},"source":[]},{"cell_type":"markdown","metadata":{"id":"cfHUUB4NmnNz"},"source":["## WHICH DND PATTERNS SERVE AS MOLECULAR CLOCKS? - PART 2/2"]},{"cell_type":"markdown","metadata":{"id":"rIjg6QS9MMKP"},"source":["Instead of creating an algorithm that uses a local initializing variable (in our case the kmers in the first DNA string), we will try our hand at Randomized Algorithms, starting with a RandomMotifSearch algorithm that will randomly select a kmer from each DNA string to initialize the motif matrix. <br>\n","\n","This process will most likey create a poor motif matrix set, however, running this algorithm many times and taking the best kmer set of motifs will yield a more accurate estimation."]},{"cell_type":"markdown","metadata":{"id":"ppFYprvayFOU"},"source":["#### RANDOMIZED MOTIF SEARCH 1.1"]},{"cell_type":"markdown","metadata":{"id":"3hspQWLAM0VA"},"source":["Below is our pseduo-code for the RandomizeMotifSearch.\n","```\n","RandomizedMotifSearch(Dna, k, t)\n","    randomly select k-mers Motifs = (Motif1, …, Motift) in each string from Dna\n","    BestMotifs ← Motifs\n","    while forever\n","        Profile ← Profile(Motifs)\n","        Motifs ← Motifs(Profile, Dna)\n","        if Score(Motifs) < Score(BestMotifs)\n","            BestMotifs ← Motifs\n","        else\n","            return BestMotifs\n","```\n","\n"]},{"cell_type":"code","execution_count":16,"metadata":{"executionInfo":{"elapsed":29,"status":"ok","timestamp":1745974286792,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"zcwAPbCymqvS"},"outputs":[],"source":["# RANDOMIZED MOTIF SEARCH\n","# INPUT: K and T integers and a space separted string DNA.\n","# OUTPUT: The \"best\" motif set from the intial random selection of t motifs.\n","\n","def RandomizedMotifSearch_old(DNA, k, t) -> list[str]:\n","  motifs = randomMotifGenerator(DNA, k)\n","  best_motifs = motifs.copy()\n","  dna_list = DNA.split()\n","  # print(motifs)\n","\n","  # while True will continue the loop indefinitely\n","  count = 0\n","  '''\n","  main issue was not having an independent outer For loop that did 1000 iterations\n","  and THEN having the interior while loop that would run indefinitel, checking for improvements on each interation\n","  of the outer loop\n","\n","  '''\n","  while count <= 1000:\n","    count += 1\n","    profile = countMotifPercentPseudocounts(motifs)\n","    # create a new set of motifs based on the profile\n","    for string in dna_list:\n","      nextMotif = ProfileMostProbableKmer(string, k, profile)\n","      motifs.append(nextMotif)\n","    # now score Motifs vs BestMotifs, replacing bestMotifs if Motifs has a lower score\n","    if computeMotifScore(motifs) < computeMotifScore(best_motifs):\n","      best_motifs = motifs.copy()\n","      motifs = []\n","\n","  return best_motifs"]},{"cell_type":"code","execution_count":17,"metadata":{"executionInfo":{"elapsed":18,"status":"ok","timestamp":1745974286793,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"vMQm1Ei7q9wq"},"outputs":[],"source":["# url = \"/content/sample_data/dataset_30307_5 (1).txt\"\n","\n","# with open(url, 'r') as file:\n","#   lines = file.readlines()\n","#   k = int(lines[0].split()[0])\n","#   t = int(lines[0].split()[1])\n","#   DNA = lines[1]\n","\n","# answer = RandomizedMotifSearch(DNA, k, t)\n","# display(\" \".join(answer))"]},{"cell_type":"code","execution_count":18,"metadata":{"executionInfo":{"elapsed":35,"status":"ok","timestamp":1745974286827,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"DQtOMNXXSEy3"},"outputs":[],"source":["# RANDOM MOTIF GENERATOR\n","# INPUT: K integer for the lengther of the motif and a space separated string DNA\n","# OUTPUT: A randomly selected kmer from each string in DNA to create a motif list\n","\n","import random\n","\n","def randomMotifGenerator(DNA, k) -> list[str]:\n","  motifs = []\n","  dna_list = DNA.split()\n","  for string in dna_list:\n","    random_index = random.randint(0, len(string)-k)\n","    motif = string[random_index:random_index+k]\n","    motifs.append(motif)\n","  return motifs\n","\n","# HAMMING DISTANCE\n","# INPUT: Two strings of equal length\n","# OUTPUT: The hamming distance between the two strings\n","\n","def HammingDistance(p: str, q:str) -> int:\n","  hd = 0\n","  for i in range(len(p)):\n","    if p[i] != q[i]:\n","      hd += 1\n","  return hd\n"]},{"cell_type":"code","execution_count":19,"metadata":{"executionInfo":{"elapsed":3,"status":"ok","timestamp":1745974286828,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"pOJa11ZsPb3O"},"outputs":[],"source":["# -- UPDATED VERSION -- 03/25/25\n","# ProfileMostProbableKmer -> We will find a Profile-Most probable kmer of a string\n","# INPUT: a string TEXT, integer k, and a 4xk profile matrix (row x column)\n","# OUTPUT: A Profile-most probable kmer in TEXT.\n","\n","# we are basing out rows from previous examples, order is as follows:\n","# 0 -> A, 1 -> C, 2 -> G, 3 -> T\n","def ProfileMostProbableKmer(text: str, k: int, profiles) -> str:\n","  mostProbableKmer = text[:k]\n","  kmers = set()\n","  prob_table = {}\n","  # lets list our kmers\n","  for i in range(len(text)-k+1):\n","    kmer = text[i:i+k]\n","    kmers.add(kmer)\n","\n","  # for each kmer, we want to compute its probability based on the given probabilty matrix\n","  for kmer in kmers:\n","    # setting the base probability as 1, but this will change accordingly when multiplied by the actual probability.\n","    prob_table[kmer] = 1\n","\n","    # for each nucleotide in kmer, we check which it is and then multiply the appropriate probability based on its position in the string.\n","    for column in range(k):\n","      # UPDATE\n","      # using the .get() method allows us to only obtain the probability in profile correspondin to the nucleotide and position of kmer[column]\n","      prob_table[kmer] = (prob_table[kmer] * profiles[column].get(kmer[column],0))\n","\n","  # now we look for the max probability value in our table\n","  max_val = max(prob_table.values())\n","\n","  # now we look for the kmer that has that max value\n","  for kmer in prob_table:\n","    if max_val == 0:\n","      return mostProbableKmer\n","\n","    if prob_table[kmer] == max_val:\n","      mostProbableKmer = kmer\n","\n","  return mostProbableKmer\n","\n","\n","# countMotifPercent\n","# INPUT: A list of DNA strings\n","# OUTPUT: A profile matrix, showing the probability of a nucleotide being present at different indices\n","\n","def countMotifPercent(motifs):\n","    count = {}\n","    columns = []\n","    for i in range(len(motifs[0])):\n","        columns.append([motif[i] for motif in motifs])\n","    for i in range(len(columns)):\n","        count[i] = {'A': columns[i].count('A')/len(columns[i]), 'C': columns[i].count('C')/len(columns[i]), 'G': columns[i].count('G')/len(columns[i]), 'T': columns[i].count('T')/len(columns[i])}\n","\n","    return count\n"]},{"cell_type":"code","execution_count":20,"metadata":{"executionInfo":{"elapsed":2,"status":"ok","timestamp":1745974286828,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"fSw9-RX_PNKu"},"outputs":[],"source":["# CODE FROM PREVIOUS WEEK\n","# --- UPDATED VERSION ---\n","# 03/26/25\n","\n","def GreedyMotifSearch(DNA: str, k: int, t: int) -> list[str]:\n","    dnaList = DNA.split()\n","    bestMotifs = [string[:k] for string in dnaList]  # Initialize with first k-mers\n","    bestScore = float('inf')  # Track the best score (lower is better)\n","\n","    # Iterate over all possible k-mers in the first DNA string\n","    for i in range(len(dnaList[0]) - k + 1):\n","        motifs = [dnaList[0][i:i+k]]  # Start with the current k-mer from the first sequence\n","\n","        # Build the profile and find motifs in the remaining sequences\n","        for j in range(1, t):\n","            profile = countMotifPercent(motifs)\n","            nextMotif = ProfileMostProbableKmer(dnaList[j], k, profile)\n","            motifs.append(nextMotif)\n","\n","        # Compute the score of the current motif set\n","        currentScore = computeMotifScore(motifs)\n","\n","        # Update bestMotifs if the current set is better\n","        if currentScore < bestScore:\n","            bestScore = currentScore\n","            bestMotifs = motifs.copy()\n","\n","    return bestMotifs\n","\n","\n","def computeMotifScore(motifs: list[str]) -> int:\n","    \"\"\"Compute the total number of mismatches against the consensus motif.\"\"\"\n","    if not motifs:\n","        return 0\n","\n","    consensus = []\n","    k = len(motifs[0])\n","\n","    # Build the consensus string\n","    for i in range(k):\n","        column = [motif[i] for motif in motifs]\n","        mostCommon = max(set(column), key=column.count)\n","        consensus.append(mostCommon)\n","    consensusMotif = ''.join(consensus)\n","\n","    # Calculate total mismatches\n","    score = 0.0\n","    # for motif in motifs:\n","    #     score += sum(1 for a, b in zip(motif, consensusMotif) if a != b)\n","    for motif in motifs:\n","      score += HammingDistance(motif,consensusMotif)\n","\n","    return score"]},{"cell_type":"code","execution_count":21,"metadata":{"executionInfo":{"elapsed":10,"status":"ok","timestamp":1745974286836,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"edGxQbThZOg-"},"outputs":[],"source":["def countMotifPercentPseudocounts(motifs):\n","    count = {}\n","    columns = []\n","    for i in range(len(motifs[0])):\n","        columns.append([motif[i] for motif in motifs])\n","    for i in range(len(columns)):\n","        # adding a psuedocount (+1) to remove probabilities of zero\n","        count[i] = {'A': (columns[i].count('A'))+1/len(columns[i])+4, 'C': (columns[i].count('C'))+1/len(columns[i])+4, 'G': (columns[i].count('G'))+1/len(columns[i])+4, 'T': (columns[i].count('T'))+1/len(columns[i])+4}\n","\n","    return count"]},{"cell_type":"code","execution_count":22,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":11,"status":"ok","timestamp":1745974286837,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"RxojPQduaQq0","outputId":"b5027161-9a14-4446-9b19-4e4706375756"},"outputs":[{"output_type":"stream","name":"stdout","text":["{0: {'A': 5.2, 'C': 6.2, 'G': 5.2, 'T': 5.2}}\n","[5.2, 6.2, 5.2, 5.2]\n","A\n"]}],"source":["# testing the syntax to retrieve the probabilites of each\n","motif = \"ACCTG\"\n","motif_percents = countMotifPercentPseudocounts(motif)\n","print(motif_percents)\n","values = motif_percents[0].values()\n","print(list(values))\n","\n","keys = motif_percents[0].keys()\n","print(list(keys)[0])"]},{"cell_type":"code","execution_count":23,"metadata":{"executionInfo":{"elapsed":1,"status":"ok","timestamp":1745974286886,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"DYLrXy87sm_a"},"outputs":[],"source":["# --- UPDATED --- 04/03/2025\n","\n","def RandomizedMotifSearch(DNA, k, t) -> list[str]:\n","    dna_list = DNA.split()\n","    best_motifs = []\n","    best_score = float('inf')  # Initialize with worst possible score\n","\n","    for _ in range(1000):  # Run 1000 independent restarts\n","        # 1) Generate random starting motifs\n","        motifs = randomMotifGenerator(DNA, k)\n","        current_score = computeMotifScore(motifs)\n","\n","        while True:  # Iterate until no improvement\n","            # 2) Build profile from current motifs\n","            profile = countMotifPercentPseudocounts(motifs)\n","\n","            # 3) Generate new motifs using the profile\n","            new_motifs = []\n","            for string in dna_list:\n","                nextMotif = ProfileMostProbableKmer(string, k, profile)\n","                new_motifs.append(nextMotif)\n","\n","            # 4) Check if new motifs are better\n","            new_score = computeMotifScore(new_motifs)\n","\n","            if new_score < current_score:\n","                motifs = new_motifs\n","                current_score = new_score\n","            else:\n","                break  # No improvement, exit loop\n","\n","        # 5) Update best_motifs if this run found a better set\n","        if current_score < best_score:\n","            best_motifs = motifs.copy()\n","            best_score = current_score\n","\n","    return best_motifs"]},{"cell_type":"code","source":["DNA = \"TGACGTTC TAAGAGTT GGACGAAA  CTGTTCGC\"\n","motifs = [\"TGA\",\"GTT\",\"GAA\",\"TGT\"]\n","\n","dna_strings = DNA.split()\n","profile = countMotifPercentPseudocounts(motifs)\n","\n","new_motifs = []\n","for string in dna_strings:\n","  nextMotif = ProfileMostProbableKmer(string, 3, profile)\n","  new_motifs.append(nextMotif)\n","\n","print(new_motifs)\n","display(\" \".join(new_motifs))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":53},"id":"B5XVEEP3LcyG","executionInfo":{"status":"ok","timestamp":1745974601167,"user_tz":240,"elapsed":8,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"c0ea90c0-0c8a-45fa-880e-4570466ea33b"},"execution_count":37,"outputs":[{"output_type":"stream","name":"stdout","text":["['TGA', 'TAA', 'GGA', 'TGT']\n"]},{"output_type":"display_data","data":{"text/plain":["'TGA TAA GGA TGT'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{}}]},{"cell_type":"code","execution_count":24,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":705,"status":"ok","timestamp":1745974287591,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"HEz0A0uqX2m5","outputId":"721c1989-7fc3-433b-b6d1-0cf426fc7d20"},"outputs":[{"output_type":"stream","name":"stdout","text":["['TCTCGGGG', 'CCAAGGTG', 'TACAGGCG', 'TTCAGGTG', 'TCCACGTG']\n"]}],"source":["DNA = \"CGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG TAGTACCGAGACCGAAAGAAGTATACAGGCGT TAGATCAAGTTTCAGGTGCACGTCGGTGAACC AATCCACCAGCTCCACGTGCAATGTTGGCCTA\"\n","k = 8\n","t = 5\n","\n","print(RandomizedMotifSearch(DNA, k, t))"]},{"cell_type":"markdown","metadata":{"id":"GQfejgdzx0B-"},"source":["RandomizedMotifSearch has a large beenift of being able to run longer motifs since it has a lower time complexity than the GreedyMotifSearch."]},{"cell_type":"markdown","metadata":{"id":"t3iik67Cx-k1"},"source":["#### GIBB'S SAMPLING 1.3"]},{"cell_type":"markdown","metadata":{"id":"a-T0mBWya7mi"},"source":["Like RandomizedMotifSearch, GibbsSampler starts with randomly chosen k-mers in each of t DNA sequences, but it makes a random rather than a deterministic choice at each iteration. It uses randomly selected k-mers (Motif1, …, Motift) to come up with another (hopefully better scoring) set of k-mers. In contrast with RandomizedMotifSearch (which deterministically defines new motifs as <br><br>\n","\n","Motifs(Profile(Motifs), Dna), <br><br>\n","\n","GibbsSampler randomly selects an integer i between 1 and t and then randomly changes only a single k-mer Motifi."]},{"cell_type":"markdown","metadata":{"id":"EfpmFrGPrkDj"},"source":["Creating a function RandomGenerator() that takes the probabilities given and divides each by the sum of the probabilities. Since this will be a non-negative set of probabilities, which could not sum up to 1."]},{"cell_type":"code","execution_count":25,"metadata":{"executionInfo":{"elapsed":1,"status":"ok","timestamp":1745974287592,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"5ICeiTavx9Oi"},"outputs":[],"source":["import random\n","# INPUT: A list of non-negative probabilites\n","# OUTPUT: A list of probabilites that sum to 1\n","def GibbsRandomGenerator(probabilities):\n","  # need to return probabilities that are each divided by the sum\n","  # WAS RUNNING TOO SLOW\n","  new_probabilities = probabilities.copy()\n","  for i in range(len(probabilities)):\n","    new_probabilities[i] = new_probabilities[i]/sum(probabilities)\n","  # new_probabilities = new_probabilities/probabilities.sum()\n","\n","  return new_probabilities"]},{"cell_type":"code","execution_count":26,"metadata":{"executionInfo":{"elapsed":7,"status":"ok","timestamp":1745974287600,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"tsD0OmhktEAw"},"outputs":[],"source":["# INPUT: A list of non-negative probabilites\n","# OUTPUT: An index chosen by weighted probabilites of each index\n","def ChooseWeightedRandom(probabilities):\n","  # given numbers are a set of non-zero probabilites, lets make the set equal to 1:\n","  weighted_set = GibbsRandomGenerator(probabilities)\n","\n","  # now that we have a weighted set, we choose an index based on the probability of the index\n","  random_index = random.choices(range(len(weighted_set)), weights=weighted_set, k=1)\n","  return random_index[0]"]},{"cell_type":"code","execution_count":27,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1745974287603,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"_CJM6et_vKtp","outputId":"d5ffa715-62d4-48d7-d23d-5d2c3e176e84"},"outputs":[{"output_type":"stream","name":"stdout","text":["2\n","1\n","2\n","1\n","2\n","1\n","1\n","1\n","2\n","2\n","2\n"]}],"source":["probs = [0.1, 0.2, 0.3]\n","n = 0\n","while n < 10:\n","  print(ChooseWeightedRandom(probs))\n","  n += 1\n","print(ChooseWeightedRandom(probs))"]},{"cell_type":"markdown","metadata":{"id":"keZOm4duFLeE"},"source":["\"GibbsSampler uses this random number generator to select a Profile-randomly generated k-mer at each step: if the die rolls the number i, then we define the Profile-randomly generated k-mer as the i-th k-mer in Text. While the pseudocode below repeats this procedure N times, in practice GibbsSampler depends on various stopping rules that are beyond the scope of this chapter\" - *Bioinformatic Algorithms*."]},{"cell_type":"markdown","metadata":{"id":"rDuC-iA0jMJc"},"source":["Below will represent our pseudocode for the Gibbs Sampler function.\n","\n","```\n","GibbsSampler(Dna, k, t, N)\n","    randomly select k-mers Motifs = (Motif1, …, Motift) in each string from Dna\n","    BestMotifs ← Motifs\n","    for j ← 1 to N\n","        i ← Random(t)\n","        Profile ← profile matrix constructed from all strings in Motifs except for Motifi\n","        Motifi ← Profile-randomly generated k-mer in the i-th sequence\n","        if Score(Motifs) < Score(BestMotifs)\n","            BestMotifs ← Motifs\n","    return BestMotifs\n","```\n","\n"]},{"cell_type":"code","execution_count":28,"metadata":{"executionInfo":{"elapsed":4,"status":"ok","timestamp":1745974287606,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"LnYx7Oa8E4yk"},"outputs":[],"source":["# GIBBS SAMPLER\n","# INPUT: Integers k, t, and N representing the length of our k-mers, number of DNA strings and number of times to run GIBBS SAMPLER respectively.\n","#     followed by space separated string DNA.\n","# OUTPUT: Best Motifs for the selected DNA strings, but this time randomly changing one kmer selection instead of all of them at once.\n","\n","def GibbsSampler(k, t, N, DNA) -> list[str]:\n","  # 1) Randomly select a set of motifs from each DNA substring\n","  motifs = randomMotifGenerator(DNA, k)\n","  best_motifs = motifs.copy()\n","  dna_list = DNA.split()\n","\n","  for _ in range(N): # this loop will run for N iterations\n","    # 2) randomly select an index(i) in order to remove motif(i) from our collection of motifs\n","    random_index = random.randint(0, t-1)\n","    dna_substring = dna_list[random_index]\n","\n","    # 3) Now compute the profile matrix using the set of motifs, excluding the randomely selected one\n","    # ISSUE: If the random_index = k, then motifs[k+1] will return only the first motif as it overflows\n","    current_motifs = motifs.copy()\n","    removed_motif = current_motifs.pop(random_index)\n","    profile = countMotifPercentPseudocounts(motifs)\n","    '''\n","    THE NESTED FOR LOOP BELOW IS CAUSING SERIOUS TIME CONCERNS 04/17/2025\n","    '''\n","    # 4) Compute the probability of each motif in the removed DNA substring i; given our computed Profile Matrix\n","    kmers = []\n","    prob_table = {}\n","    for i in range(len(dna_substring)-k+1):\n","      kmer = dna_substring[i:i+k]\n","      kmers.append(kmer)\n","      # for each kmer, we want to know the probability basd on the givien profile matrix\n","      for kmer in kmers:\n","        prob_table[kmer] = 1\n","\n","        # for each nucleotide in kmer, we check which it is and multiply the appropriate probability to the overall kmer probability\n","        for column in range(k):\n","        # using the .get() method allows us to only obtain the probability in profile correspondin to the nucleotide and position of kmer[column]\n","          prob_table[kmer] = prob_table[kmer] * profile[column].get(kmer[column],0)\n","\n","      # 5) use weighted random selection to choose which kmer to add to our motifs\n","      probabilities = list(prob_table.values())\n","      weighted_random_index = ChooseWeightedRandom(probabilities)\n","\n","      # now grabbing the kmer from the weighted-randomly index selection\n","      new_motif = kmers[weighted_random_index]\n","\n","      # reinsert new kmer from previously removed dna_substring\n","      # FOUND THE PROBLEM -> I popped the kmer from motifs before, but here when I attmped to splice\n","      # motifs, the kmer with the random index was not here, so with each iteration the number of kmers was decreasing\n","      # FIX: created a working copy called \"current_motifs\" that is updated at the end of each iteration\n","      # BEFORE: motifs = motifs[:random_index] + [new_motif] + motifs[random_index+1:]\n","      current_motifs = motifs[:random_index] + [new_motif] + motifs[random_index+1:]\n","      # 6) now we check if the new set of motifs have a better motif score then the current \"best\" set of motifs\n","    # updating 'motifs' to be equal to 'current_motifs'\n","    motifs = current_motifs.copy()\n","    if computeMotifScore(motifs) < computeMotifScore(best_motifs):\n","      best_motifs = motifs.copy()\n","\n","      # 7) repeat N times\n","\n","  return best_motifs\n","\n","\n","\n"]},{"cell_type":"code","execution_count":33,"metadata":{"executionInfo":{"elapsed":22,"status":"ok","timestamp":1745974308797,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"WW0VPNP8k49M"},"outputs":[],"source":["# k = 8\n","# t = 5\n","# N = 100\n","# DNA = \"CGCCCCTCTCGGGGGTGTTCAGTAACCGGCCA GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG TAGTACCGAGACCGAAAGAAGTATACAGGCGT TAGATCAAGTTTCAGGTGCACGTCGGTGAACC AATCCACCAGCTCCACGTGCAATGTTGGCCTA\"\n","\n","# answer = GibbsSampler2(k,t,N,DNA)\n","# print(answer)\n","# display(\" \".join(answer))\n","# 04/17/25 : Getting an out of bounds warning for length of motifs - FIXED on 04/17/25\n","# 04/17/25 : Getting the correct number of BestMotifs, but the incorrect kmers for BestMotifs - ERROR"]},{"cell_type":"code","execution_count":32,"metadata":{"executionInfo":{"elapsed":5,"status":"ok","timestamp":1745974298097,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"6L1KOS-OeNu_"},"outputs":[],"source":["def computeMotifScore2(motifs: list[str]) -> int:\n","    \"\"\"Compute the total mismatches against the consensus, using pseudocounts.\"\"\"\n","    if not motifs:\n","        return 0\n","\n","    k = len(motifs[0])\n","    consensus = []\n","    bases = {'A', 'C', 'G', 'T'}\n","\n","    for i in range(k):\n","        counts = {'A': 1, 'C': 1, 'G': 1, 'T': 1}  # Pseudocounts\n","        for motif in motifs:\n","            counts[motif[i]] += 1\n","        consensus.append(max(counts.items(), key=lambda x: x[1])[0])\n","\n","    consensusMotif = ''.join(consensus)\n","    return sum(HammingDistance(motif, consensusMotif) for motif in motifs)"]},{"cell_type":"code","execution_count":30,"metadata":{"executionInfo":{"elapsed":1,"status":"ok","timestamp":1745974293152,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"WPX0kJo7-ITn"},"outputs":[],"source":["from types import new_class\n","import random\n","import numpy as np\n","\n","def GibbsSampler2(k, t, N, DNA) -> list[str]:\n","  for _ in range(20):\n","      dna_list = DNA.split()\n","      motifs = randomMotifGenerator(DNA, k)\n","      best_motifs = motifs.copy()\n","      best_score = computeMotifScore(best_motifs)\n","\n","      for _ in range(N):\n","          # Randomly select a sequence to exclude\n","          random_index = random.randint(0, t-1)\n","          excluded_seq = dna_list[random_index]\n","\n","          # Create profile from remaining motifs\n","          profile = countMotifPercentPseudocounts([motif for i, motif in enumerate(motifs) if i != random_index])\n","\n","          # Compute all kmer probabilities for the excluded sequence GIVEN the probability matrix \"profile\".\n","          kmers = [excluded_seq[i:i+k] for i in range(len(excluded_seq)-k+1)]\n","          probabilities = np.zeros(len(kmers))\n","\n","          # Calculate probabilities\n","          for i, kmer in enumerate(kmers):\n","              prob = 1.0\n","              for pos in range(k):\n","                  prob *= profile[pos].get(kmer[pos], 1)\n","              # the index in probabilities corresponds to the index in kmers\n","              probabilities[i] = prob\n","\n","          # Normalize probabilities and select new motif from excluded set using GibbsRandomGenerator\n","          # normalized_probabilities = GibbsRandomGenerator(probabilities)\n","          selected_idx = ChooseWeightedRandom(probabilities)\n","          new_motif = kmers[selected_idx]\n","\n","\n","          # Update motif at the previously randomly selected index\n","          motifs[random_index] = new_motif\n","\n","          # Update best motifs if improved\n","          current_score = computeMotifScore2(motifs)\n","          if float(current_score) < float(best_score):\n","              best_motifs = motifs.copy()\n","              best_score = current_score\n","          # if N % 10 == 0:  # Print every 10th iteration\n","          # # SCORE NOT IMPROVING!!\n","          #     print(f\"Iteration {_}, Current Score: {current_score}, Best Score: {best_score}\")\n","\n","  return best_motifs"]},{"cell_type":"markdown","metadata":{"id":"7rFTVcvanSPy"},"source":["As of 04/24/25 the GibbsSampler2 is not returning the expected strings."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":41,"status":"aborted","timestamp":1745974287666,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"},"user_tz":240},"id":"rh50sir4psJP"},"outputs":[],"source":["url = \"/content/sample_data/dataset_30309_11 (3).txt\"\n","\n","with open(url, 'r') as file:\n","  lines = file.readlines()\n","  k = int(lines[0].split()[0])\n","  t = int(lines[0].split()[1])\n","  N = int(lines[0].split()[2])\n","  DNA = lines[1]\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"k-QsAWcZv6jA","executionInfo":{"status":"aborted","timestamp":1745974287667,"user_tz":240,"elapsed":976,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"outputs":[],"source":["answer = GibbsSampler2(k,t,N,DNA)\n","display(\" \".join(answer))"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wuUoGGpTwAFX","executionInfo":{"status":"aborted","timestamp":1745974287668,"user_tz":240,"elapsed":976,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"outputs":[],"source":[]}],"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOL/nOn4wdMLgZBBWQP/QRW"},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}