{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMV1nj5Gp4XLH5gXYldfe4n"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["## WHICH DND PATTERNS SERVE AS MOLECULAR CLOCKS? - PART 1/2"],"metadata":{"id":"CFOk6XVPZQui"}},{"cell_type":"markdown","source":["### DO WE HAVE A \"CLOCK\" GENE? 1.1"],"metadata":{"id":"FJKx_t3DZxYO"}},{"cell_type":"markdown","source":["We will turn our focus to the internal time keeper for plants, animals ans some bacteria - the **Circadian Clock**. The genes that are involved in letting our body know what \"time\" it is in order to change their gene transpcrition prduction, or **gene expression**. <br>\n","\n","For plants, each cell keeps track of the day and night cycle independently using three plant genes : LHY, CCA1, and TOC1."],"metadata":{"id":"tc36OZPXZ5Tm"}},{"cell_type":"markdown","source":["Theses are regulatory genes that encode for **regulatory proteins** that are usually impacted by external factors (such as the presense of certain nutrients or even sunlight). For these genes, TOC1 promotes the expression of LHY and CCA1, while LHY and CCA1 repress the expression of TOC1, resulting in a **negative feedback loop**. <br><br>\n","In the morning, the sunlight activates the transcription of LHY and CCA1, triggering the repression of TOC1 transcription. When the sunlight diminishes, the transcription of LHY and CCA1 diminishes as well, causing TOC1 to no longer be repressed. TOC1 peaks at night and starts promoting LHY and CCA1, continuing the cycle."],"metadata":{"id":"fWsPZRwJ2jN4"}},{"cell_type":"markdown","source":["The regulatory protiens that these genes encode are known as **transcritpion factors** that can turn other genes on or off. A transcription factor regulates a gene by binding to specific short DNA intervals known as **regulatory motifs** in the gene's upstream region."],"metadata":{"id":"w851wvFC35ni"}},{"cell_type":"markdown","source":["### MOTIF FINDING IS MORE DIFFICULT THAN YOU THINK 1.2"],"metadata":{"id":"HHvIoA70T7Ud"}},{"cell_type":"markdown","source":["In order to find motifs we may think to use our earlier \"Frequent Words\" algorithm, but the Frequent Words Problem is inadequate because it does not correctly model the biological problem of motif finding. <br> <br>\n","A DnaA box is a pattern that clumps, or appears frequently, within a relatively short interval of the genome. In contrast, a regulatory motif is a pattern that appears at least once (perhaps with variation) in each of many different regions that are scattered throughout the genome."],"metadata":{"id":"hpJTIHsCVnp9"}},{"cell_type":"markdown","source":["Let us work out a \"brute force\" method for finding these motifs. <br>\n","Given a collection of strings Dna and an integer d, a k-mer is a (k,d)-motif if it appears in every string from Dna with at most d mismatches. For example, the implanted 15-mer in the strings above represents a (15,4)-motif."],"metadata":{"id":"MqBTTz3-V9n2"}},{"cell_type":"markdown","source":["```\n","# This is formatted as code\n","MotifEnumeration(Dna, k, d)\n","    Patterns ← an empty set\n","    for each k-mer Pattern in the first string in Dna\n","        for each k-mer Pattern’ differing from Pattern by at most d mismatches\n","            if Pattern' appears in each string from Dna with at most d\n","            mismatches\n","                add Pattern' to Patterns\n","    remove duplicates from Patterns\n","    return Patterns\n","```\n","\n"],"metadata":{"id":"498knl04WonR"}},{"cell_type":"code","source":["# # --- OLD VERSION ---\n","# # Implemented Motif Problem\n","# # INPUT: A collection of strings \"DNA\" and integers k and d.\n","# # OUTPUT: All (k,d)-motifs in DNA\n","# def MotifEnumeration(DNA: str, k:int, d:int) -> list[str]:\n","#   patterns = set()\n","#   first_patterns = set()\n","#   n = len(DNA[0])\n","#   first_string = DNA[0]\n","\n","#   # we will create a list of patterns found in the first string of DNA\n","#   for i in range(n-k+1):\n","#     pattern = first_string[i:i+k]\n","#     first_patterns.add(pattern)\n","\n","#   # now we will create a neighborhood for each pattern having at most d mismatches\n","#   neighborhood = []\n","#   # print(first_patterns) ->  WORKING\n","\n","#   for pattern in first_patterns:\n","#     neighbors = Neighbors(pattern,d)\n","#     neighborhood.update(neighbors)\n","#     # checking if each neighbor is present in all of the other strings in DNA\n","#   for neighbor in neighborhood:\n","#     # PROBLEM here is that we are checking for EXACT matches in the other DNA strings\n","#     if all(neighbor in dna_substring for dna_substring in DNA[1:]) == True:\n","#       patterns.append(neighbor)\n","\n","\n","#   return patterns\n","\n"],"metadata":{"id":"31hYsXM92h9X"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# --- OLD VERSION ---\n","# Implemented Motif Problem\n","# INPUT: A collection of strings \"DNA\" and integers k and d.\n","# OUTPUT: All (k,d)-motifs in DNA\n","def MotifEnumeration(DNA: str, k:int, d:int) -> list[str]:\n","  patterns = set()\n","  first_patterns = set()\n","\n","  if isinstance(DNA,list):\n","    n = len(DNA[0])\n","  else:\n","    DNA = DNA.split()\n","    n = len(DNA[0])\n","\n","  first_string = DNA[0]\n","\n","  # we will create a list of patterns found in the first string of DNA\n","  for i in range(n-k+1):\n","    pattern = first_string[i:i+k]\n","    first_patterns.add(pattern)\n","\n","  # now we will create a neighborhood for each pattern having at most d mismatches\n","  neighborhood = set()\n","  # print(first_patterns) ->  WORKING\n","\n","  for pattern in first_patterns:\n","    neighbors = Neighbors(pattern,d)\n","    neighborhood.update(neighbors)\n","    # checking if each neighbor is present in all of the other strings in DNA\n","  for neighbor in neighborhood:\n","    isMotif = True\n","    for dna_string in DNA[1:]:\n","      found = False\n","      for i in range(len(dna_string)-k+1):\n","        kmer = dna_string[i:i+k]\n","        if HammingDistance(kmer,neighbor) <= d:\n","          found = True\n","          # found that the neighbor is in this dna_substring\n","          # exiting this for loop and moving on to next substring\n","          break\n","      if not found:\n","        # not in this substring, no longer qualifies for motif, next neighbor\n","        isMotif = False\n","        break\n","    if isMotif:\n","      # kmer was found in each DNA substring, can now add it to patterns\n","      patterns.add(neighbor)\n","\n","  return patterns"],"metadata":{"id":"t8qQBDlBmTIH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# HAMMING DISTANCE\n","# INPUT: Two strings of equal length\n","# OUTPUT: The hamming distance between the two strings\n","\n","def HammingDistance(p: str, q:str) -> int:\n","  hd = 0\n","  for i in range(len(p)):\n","    if p[i] != q[i]:\n","      hd += 1\n","  return hd\n","\n","\n","# first we need to make our Neighborhood function.\n","# NEIGHBORS FUNCTION\n","# Recursive Function that takes an input pattern (string) and outputs a collection of neighbor sequences that have a hamming distance of at most d.\n","# INPUT: A string Pattern and an integer d (for the max hamming distance)\n","# OUTPUT: A collection of strings Neighbors(pattern,d)\n","\n","def Neighbors(pattern: str, d: int) -> list[str]:\n","  nucleotides = ['G', 'T', 'A','C']\n","\n","  # for HammingDistance to be 0, neighbor == pattern\n","  if d == 0:\n","    return [pattern]\n","  # If length pattern = 1, then it is a single nucleotide, so the neightbors will just be the other nucleotides\n","  if len(pattern) == 1:\n","    return nucleotides\n","  neighborhood = [] # creating empty set\n","  SuffixPattern = pattern[1:] # Dropping the first symbol in pattern to get the SuffixPattern\n","  SuffixNeighbors = Neighbors(SuffixPattern, d) # Recursion, will keep calling function until base case (len(pattern) == 1)\n","\n","  for text in SuffixNeighbors:\n","    if HammingDistance(SuffixPattern, text) < d: # <d = there can be more mismatching allowed\n","      for nucleotide in nucleotides:\n","        # we now add a nucleotide as each possible neigbor to pattern\n","        neighbor = nucleotide + text\n","        neighborhood.append(neighbor)\n","    else: # HammingDistance is equal to d, thus no more mismatching allowed and the prefix needs to be the prefix of the original pattern\n","      neighbor = pattern[0] + text # SuffixPattern[0] is the first symbol in SuffixPattern\n","      neighborhood.append(neighbor)\n","\n","  return neighborhood"],"metadata":{"id":"jU-fAYB9ZiJZ"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# testing MOTIF ENUMERATION function\n","DNA = \"ATTTGGC TGCCTTA CGGTATC GAAAATT\"\n","DNA_list = DNA.split()\n","k = 3\n","d = 1\n","\n","answer = MotifEnumeration(DNA, k, d)\n","print(\"your motifs are:\",answer)\n","# CORRECT"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PbC1_VVqZnKF","executionInfo":{"status":"ok","timestamp":1743023161328,"user_tz":240,"elapsed":152,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"b3f0d885-c625-4bf8-cb43-a8a6ac96f9fe"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["your motifs are: {'GTT', 'ATT', 'ATA', 'TTT'}\n"]}]},{"cell_type":"code","source":["if isinstance(DNA,list):\n","  print(\"is a list\")\n","else:\n","  print(\"is not a list\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"x8E18sk3HBNc","executionInfo":{"status":"ok","timestamp":1743023161330,"user_tz":240,"elapsed":2,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"de7f50e5-ee1e-4beb-9e62-950faada1795"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["is not a list\n"]}]},{"cell_type":"code","source":["# test data set given from Coursera\n","DNA = \"GATGGGCGAACTTCAGCGCGGCCGA TCAGACTATCGCGCGCGACATGGAC GCGAACCGCTTGTTTACTGACCGAC CGTTGGCAGAAGATTGCGCCTAGTT AAGCCGGGGCATAGTGCGTCTGTCC ATTGTGCGGTGACTTCTCAAGGGTC\"\n","k = 5\n","d = 2\n","\n","answer = MotifEnumeration(DNA, k,d)\n","display(\" \".join(answer))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":174},"id":"4vsEoXs3HLl8","executionInfo":{"status":"ok","timestamp":1743023161335,"user_tz":240,"elapsed":4,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"10a1fd92-2454-4e89-88c5-d71f7c36aea5"},"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["'CCTGA CTTCC AGCTC TACGG CCTGC TTGAC TTGCC TGTAG CTTGC TAGGG GCAGG TCGTA TTTGG ATAGA ATGAG GCGCC TGCTC CTCGC TGAGC GACCC TCTGC CCGTC ACTAT CTACG CTATA CGATA ACGGC ATAAC ATGCT AAGCG ATTAC ACGTA CGATG CGCTC GGCCC GTACC ATACG GAGGG CAGAA AACCT TTAGC CTGCA TACCT GTTGC CCGGA CAAAC TAACG CATGG GTGAA AGTAC GTGCG GACCT TGGTG AATCT GCTAG TAAGG CTGGT ATAGC CTGGA AGGGC TATGC AGCCG GAGTG TATAG ACAGC CCGGG TATCT ACTGA TGAAC CTTTC GTAGC GACTG TGAGG CCGTG GGTCT TCCCC TTACC CCGGC CTTGT GATCG GCAAA GCGGA AGACC GGATT GGCCT ACTCC AGGAC GTCAG TGCCA TATGA TGCTA GTATT CTTAC ATGGA GTTGG TACGC CAGTT TTCGA TGAGT GACCG TTGGC GTTCG CGCTG TCCGC GCGTG CTAAG CCCGA CGGCC CGCCA CGTGA TGTTC CGCAA ATGGT CGCGA GAGAA CGCGG TTCGT TCAGC GCGAA TCGCA TACGA GCCAA GCACG TCGCC AGCGT AAAGG TGAAT ATTGT GGATC TCGAC CACAT CACTG AGCTT CGTCT TCCGT CTAGC TATGG TGGTA ACCAG CGCTT GGAAA ACGAG CCGCG ATCAG GCTAA TGCGA GCCCA GTCTT TCACA GCAAG GCGTT ATCGC TGCCC GTCGT TGTGC TTCTA GTTTC GATCT TCCTG TAAGT GTCCG GCCTA CGGAA GAACG ATCCT CATCT GCCAC GGTGA GAATA GACGT ACGCA AAGGC GGCTT GCCCG GATAC CGAGC GCTCC CTAGA CAGCT ACTGC GAGGA GGTCA GGCAT GGTGG CATTG CGAGT AAGCA GCGAC GTGTG CTTCG GCAGA GGCGC GCATT CTCAT GCTGA AGCCT TGATC TAGAC GCTCT TGCGT TGGAT GTAGA AGTGA TTTTG GCTTA GGGAG CGTTG GGGTA TCAGA AAAGC AGGCT GGGCG CTGCG CCCGT CCGCT AGTGG ACGTT TTGTC GAAAT TGGTT ACTTC TCAAA TACAG GGCAG AATCC TTAGT CCAAG GAGGC GAGAT CGTTC ATCTT GTGGC AGAGA CAGGC CATTT GCCAG TGTGA AATGG CCACA TAGAG GAGCA TACAT CTTAT TCGAT GGACA TGCAT GACTA CAGAG GGCTC AAGAG ACTAC TTCTC TGGCG ACATG AATGT CATGC AGATT GTCGC TCTTA GACAA AGTCA GAGCG ATGAC CGAGG GGAGC CCGCA GTCTA TGTCC ATTCT CGGAC ATACT AACAG CCTCG GGCGA CTGGC AATAG CTAAC ATTGA GCGAG ACGAC GGTAC TACGT TACTT CAAAA CGGCG TCACC GCCTT GGCGT GCGGT GACCA TCTGA CAGTC GAATT CTCCC ATTGG CGATT CAGAC TACCG ATCCC TCCCG GCTTC GCACA AAACG AGGCC TGACC ATCGG TTCCG ACATC CGTAG GCCTC TTTGA CATGA CTCGG TACTA TGGCT GATAA TCCGG TCTCA GAGAC CTTTG GCCTG AGCAA CGACG GCGAT GGAAT CGCGC GTGGA CTTGG CGACA CGAAT TGGGC GTGAC TTTTC TCTCG GACGC CCCGG AAGCT CCAGC GCTGC TGCCT GGTTA AGTTT GTTTA GGATG TAAAG GACAT GTGCT GGGCC TGCAC CCACT CTGTG GGCCG GCCGG TCAGT TTCGC GGACT TCGGC TGCGC AATTG GCACT AGCGA ATGGC GCATG CTGTA TCGCG GAGCT GTCAA CGCGT TTTGC GTCCT GACGG ATTTG CAACT GCCGT CTTTA CCGAC ACCTG AGAAG GCCGC GGCTG CCCTA AAGGA CGCAT GCTCA AATGA CTAAT CTAGG TTACG ATATT GCGTC ACGCC ATCTC GCTGT ACTGT CCAGT TCGGT GTCGA TGCAG TGACG CCAAA TTTAG GATGT ACCTC ACGCG CGTCC TGTAC TGAGA TTGTG GTGGT GACGA TGGTC CCGAT CGGCT ACACT CGTGG CGACC GCTTG CGTTA TAGCT TTATC GAGCC ACCCC CCTCA ACGGA CACCT GATGC GCAAC TCTTT TATTC GTCCC GAGAG TAATC CGGAG GTACA AACGT TGTGG GTCGG GCTGG AGTTG ATCGA CACAC GCGGC ATTAG ACAGG TATTT GATGG GAAAG GCAGC TTACT TGCCG ACATA AGAAT CTACT TCTTG AACAT TGTCA CCTGG TAGCG CAATG GCCCT TCGTC CCTTC ACTTT ACACG GTACT TCGCT GCGTA GAGTA AGATC AGGCG GAGTC GGCAC TCTAC CGCAG TCCAG TACTC CCAGA ACGGT CCCTG TCGGA AGTGC CGGGT GGGGA TTATG ATTTC CTTTT AGAGC AAATC CGTAT TACTG GACTT GAATG GCAAT TCCGA GCATC AAATG CTGAC ACCGC CGAGA GAAGG CGATC AAACT CTCTA ATGAT GTGTA TTAGG GTGAT CAAGT TCTCC AAGTG CCCTC ACTGG CGGCA CTACC GTGCA GATTC CCACG CGTGT AGCAT CATGT GTGCC GGTCG TTGCG TAGTG GGGAT ACACC CGAAG TGGCC ATCCG AATGC CTCTG GCGCT CGGAT ATTCG CTGAA GGCTA TCATT GGACG AAGAA TTCGG AATCG GGCGG CGGGC ATGCC AACTG AAGGG CCTCT TTTAC CCGAA GCCAT CTGAG TCGTG GATTG AATAC TATGT ACAGT AGCTG GGGAA CTCTT GGCCA CAGCG GGCAA ACTAG CCTAA TGGGT ATGCG CGTAC TTGGT ACTCG GCTAC TGTCG TTTCC ACGTC AGCGG CGCCG GACAG CACCG AACGC GCGCA ACGCT GGACC TGGAC CGCAC ACAAA TTGCA ACCTA TATCG GTCCA AGCGC CTGCC TCGAG CACGT GTCTG CGACT TGACT CGGTA TTTCG TAGTC GAAGT TTGAG CGTGC ATTTT GCCGA ATCGT CGCCT CAAAT GACTC GCGGG GATCA GATTA CCTTT CCTAC GGGCA TGGCA CTTAG TCCTA GCTAT TTGGG TATTG ATTCC ACTTA GGGTG GAATC CAGGT ATTGC AGTCC GGGAC CATCG GTCAC AAGCC CGTAA GTACG TCTGG GTGTC AGGGA GGATA GCGCG CTATC GTAAT CAGTG GAGTT TTCCC GGGCT GCTCG AGGAA TGCGG TTTCT CCGAG CGGTC CCTTG CAAAG CTGGG GGAAC ATCTA CACAG ATGTC ATAGT CTCGT GAAGA ATGTT ATGTG GTGAG GCTTT TCTAG AGTGT AGACT TCTAA AGACG CTGTC TTTCA'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{}}]},{"cell_type":"markdown","source":["Although out MotifEnumeration function seems to be working, the overall time complexity is exponential. The major subfunctions in our major functions take into account the length of the target pattern sequence, possibly checknig a neighbor for each position with 4 possible nucleotide variations, as well as checking several strips of DNA sequences and checking if our pattern is present.<br><br>\n","The longer the DNA target pattern (k) and the greater number of allowable mismathces (d), our function will exponentially increase the time it will take to run. One of the major downfalls of this **\"Brute Force\"** approach."],"metadata":{"id":"G_biRdPuK74g"}},{"cell_type":"markdown","source":["#### SCORING MOTIFS 1.3"],"metadata":{"id":"w0ZPUcDJWuC_"}},{"cell_type":"markdown","source":["An inherit issue our function has when it comes to real world data is confronting noise. It is normal to find sequences that have nothing to do with let's say the circadian clock gene to be amplified in the evening.For such noisy datasets, any algorithm for the Implanted Motif Problem would fail, because as long as a single sequence does not contain the transcription factor binding site, a (k, d)-motif does not exist!"],"metadata":{"id":"8IIwSggEW1mg"}},{"cell_type":"markdown","source":["We should instead look for sequences that are close to the \"ideal\" defintion of a motif."],"metadata":{"id":"rsRx5UhQXTIr"}},{"cell_type":"markdown","source":["For our new approach to Motif Finding, we will be building a matrix using possible motifs in the DNA sequence. How we find these motifs will be discussed later, but what we do with them will be Motif Scoring, then Motif Counting, following by a Motif Profile and finally a Motif Consensus that will help us find a possible motif in the DNA sequence."],"metadata":{"id":"eWp1eYpfcqv7"}},{"cell_type":"markdown","source":["**MOTIF SCORING:** Completed by setting a matrix of n by k where n is the number of possible motifs and k is the length of each motif. We will then count the number of nucleotides that differ from the majority nucleotide in that column. We will add the totals up for each column and that total is the \"Motif Score\". We will aim to *minimize* this score.<br><br>\n","\n","**MOTIF COUNTING:**  We will then add up the number of times each nucleotide appears in each *column* and create a new mattrix. This new matrix will have four rows, one for each nucleotide and k columns. Each position corresponding to the number of times a specific nucleotide was found in that column.<br><br>\n","\n","**MOTIF PROFILE:** Now that we have the count for each nucleotide for the positions of the possible motifs, we can create a frequency table that has the frequency of the Ith nucleotide in the Jth column of the motifs.<br><br>\n","\n","**MOTIF CONSENSUS:** We take the nucleotide with the highest frequency in each  column and put together an ideal candidate for a regulatory motif for these regions."],"metadata":{"id":"KkbYNJUhdIHb"}},{"cell_type":"markdown","source":["Now there is also the issue of making sure account for more \"conservative\" locations in the motif sequence. For instance, having a column with a Motif Count of (6 C, 4 T) should be scored lower than a column of (4 C, 2 A, 2 T). The biological reason could be that a differentiation for a specific nucleotide at a certain location can still allow for protein binding, while other locations are more crucial to have the correct nucleotide for proper binding. <br><br>\n","With all this in mind we will be taking a closer look at our \"Motif Profile\" which is essentially a **Probabilty Distribution**."],"metadata":{"id":"nxfQJoedellv"}},{"cell_type":"markdown","source":["##### CLOSER LOOK AT MOTIF PROFILE AND MOTIF SCORE"],"metadata":{"id":"o8M56gDjgp0K"}},{"cell_type":"markdown","source":["**Entropy** is essentially a measure of uncertainty in a probability distribution. <br><br>\n","\n","This can be calculated by taking the negative sum of the probability of i multiplied by the Log base 2 of the probabilty of i where i starts at 0 and increments by 1 until i is equal N, the number of total items in a column.<br><br>\n","\n","We would find the entropy of more conserved columns to be lower than the entropy of the columns that are less reserved, i.e. have more variation in the probabilty distribution. <br><br>\n","\n","*Note:* Log base 2 (0) is technically undefined, but we will assume that 0 times log base 2 (0) == 0."],"metadata":{"id":"BXj7PzUhguNl"}},{"cell_type":"markdown","source":["The **Entropy** of the Motif Matrix would then be as simple as the sum of the Entropy of each column. However, for simplicity, we will continue with the Score(Motif) algorithm, knowing that an Entropy(Motif) algorithm is more often used in practice since the lower the entropy score on a column, the more conserved the column is."],"metadata":{"id":"OziCb94hj0M0"}},{"cell_type":"code","source":["# countMotifPercent\n","# INPUT: A list of DNA strings\n","# OUTPUT: A profile matrix, showing the probability of a nucleotide being present at different indices\n","\n","def countMotifPercent(motifs):\n","    count = {}\n","    columns = []\n","    for i in range(len(motifs[0])):\n","        columns.append([motif[i] for motif in motifs])\n","    for i in range(len(columns)):\n","        count[i] = {'A': columns[i].count('A')/len(columns[i]), 'C': columns[i].count('C')/len(columns[i]), 'G': columns[i].count('G')/len(columns[i]), 'T': columns[i].count('T')/len(columns[i])}\n","\n","    return count\n","\n"],"metadata":{"id":"ujp9hTvMLl8g"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["strings = \"TCGGGGGTTTTT CCGGTGACTTAC ACGGGGATTTTC TTGGGGACTTTT AAGGGGACTTCC TTGGGGACTTCC TCGGGGATTCAT TCGGGGATTCCT TAGGGGAACTAC TCGGGTATAACC\"\n","string_list = strings.split()"],"metadata":{"id":"mfwC7aLnFOxf"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["answer = countMotifPercent(string_list)"],"metadata":{"id":"zE4TziqYFwF2"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["display(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":225},"id":"qJNHR26OFw78","executionInfo":{"status":"ok","timestamp":1743023161550,"user_tz":240,"elapsed":2,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"d535e504-26df-46b5-c4cd-cfb7d53271c6"},"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["{0: {'A': 0.2, 'C': 0.1, 'G': 0.0, 'T': 0.7},\n"," 1: {'A': 0.2, 'C': 0.6, 'G': 0.0, 'T': 0.2},\n"," 2: {'A': 0.0, 'C': 0.0, 'G': 1.0, 'T': 0.0},\n"," 3: {'A': 0.0, 'C': 0.0, 'G': 1.0, 'T': 0.0},\n"," 4: {'A': 0.0, 'C': 0.0, 'G': 0.9, 'T': 0.1},\n"," 5: {'A': 0.0, 'C': 0.0, 'G': 0.9, 'T': 0.1},\n"," 6: {'A': 0.9, 'C': 0.0, 'G': 0.1, 'T': 0.0},\n"," 7: {'A': 0.1, 'C': 0.4, 'G': 0.0, 'T': 0.5},\n"," 8: {'A': 0.1, 'C': 0.1, 'G': 0.0, 'T': 0.8},\n"," 9: {'A': 0.1, 'C': 0.2, 'G': 0.0, 'T': 0.7},\n"," 10: {'A': 0.3, 'C': 0.4, 'G': 0.0, 'T': 0.3},\n"," 11: {'A': 0.0, 'C': 0.6, 'G': 0.0, 'T': 0.4}}"]},"metadata":{}}]},{"cell_type":"code","source":["import math\n","def motifEntropy(motifs):\n","    entropy = 0\n","    percents = countMotifPercent(motifs)\n","    for i in range(len(percents)):\n","        for nucleotide in percents[i]:\n","            if percents[i][nucleotide] != 0:\n","                entropy += percents[i][nucleotide] * math.log2(percents[i][nucleotide])\n","    return -entropy"],"metadata":{"id":"vmNQx22SGKoc"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["answer = motifEntropy(string_list)\n","display(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"Wtot2SuxGpHb","executionInfo":{"status":"ok","timestamp":1743023161584,"user_tz":240,"elapsed":22,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"776d05bf-bf47-4e19-b879-03f10cdd21ac"},"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["9.916290005356972"]},"metadata":{}}]},{"cell_type":"markdown","source":["#### FROM MOTIF FINDING TO FINDING A MEDIAN STRING 1.4"],"metadata":{"id":"XcPknsI9HBWH"}},{"cell_type":"markdown","source":["As stated in \"Bioinformatics Algorithms\" by Phillip Compeau and Pavel Pezner:\n","\n","> A brute force algorithm for the Motif Finding Problem (referred to as BruteForceMotifSearch) considers every possible choice of k-mers Motifs from Dna (one k-mer from each string of n nucleotides) and returns the collection Motifs having minimum score. Because there are n - k + 1 choices of k-mers in each of t sequences, there are (n - k + 1)t different ways to form Motifs. For each choice of Motifs, the algorithm calculates Score(Motifs), which requires k · t steps. Thus, assuming that k is smaller than n, the overall running time of the algorithm is O(nt · k · t). We need to come up with a faster algorithm!\n","\n"],"metadata":{"id":"NBi_THk3R0W7"}},{"cell_type":"code","source":["# MOTIF FINDING PROBLEM\n","# INPUT: A collection of strings DNA, and integer k\n","# OUTPUT: A collection of Motifs of kmers, one from each string DNA, minimizing Score(Motifs), among all possible choices of Motifs."],"metadata":{"id":"SKZb1_wyGu2y"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Being able to compute the minimum HammingDistance between a given Pattern and various patterns in a string (or row) in DNA will reduce the complexity of our algorithm since we do not have to try to compare the nucleotide count of each position of every string. We are now computing the minimum hamming distance between the given pattern and a k-mer in each INDIVIDUAL row of DNA strings <br><br>\n","The result of minimizing d(Pattern,DNAᵢ) will give us Motif(Pattern,Text); the sequence that gave us the minimized HammingDistance between Pattern and DNAᵢ."],"metadata":{"id":"Jz8RGgIydEbO"}},{"cell_type":"markdown","source":["To begin computing the MedianString, we need to create a function that calculates d(Pattern, DNA); i.e. the sum of all HammingDistances between Pattern and DNAᵢ."],"metadata":{"id":"Lf2QG3gnX-bZ"}},{"cell_type":"code","source":["# Finding the distance between a Pattern and a given string of DNA sequences.\n","# DistanceBetweenPatternAndStrings\n","# INPUT: A string pattern followed by a collection of space separated strings DNA\n","# OUTPUT: Distance of Pattern and DNA (d(Pattern,DNA))\n","\n","def DistanceBetweenPatternAndStrings(pattern, DNA) -> int:\n","  distance = 0\n","  k = len(pattern)\n","  DNA_string = DNA.split(\" \")\n","\n","  for string in DNA_string:\n","    # we are looking to MINIMIZE hd for pattern and pattern' in DNA[i]\n","    # therefore, we will set hd to infinite (or simply k, since the max hd is\n","    # equal to the length of the pattenr), and only add hd to distance that are\n","    # smaller than the previous hd\n","\n","    # resetting hd to be equal to k\n","    champion_hd = k\n","    n = len(string)\n","\n","    for i in range(n-k+1):\n","      # looking at each pattern_prime from i to n-k\n","      pattern_prime = string[i:i+k]\n","      challenger_hd = HammingDistance(pattern,pattern_prime)\n","      # we want the champion_hd to be the minimal hd between pattern and\n","      # a pattern_prime in string\n","      if champion_hd > challenger_hd:\n","        champion_hd = challenger_hd\n","\n","    # when we exit the nested for loop, we should have the minimum hd\n","    # between pattern and a given pattern within the string be\n","    # equal to champion hd, we can now add that minimized value to\n","    # the total distance\n","\n","    distance += champion_hd\n","\n","  return distance"],"metadata":{"id":"HEEqnd74qpmR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# TESTING DistanceBetweenPatternAndStrings\n","\n","pattern = \"AAA\"\n","DNA = \"TTACCTTAAC GATATCTGTC ACGGCGTTCG CCCTAAAGAG CGTCAGAGGT\"\n","\n","answer = DistanceBetweenPatternAndStrings(pattern,DNA)\n","\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"2tZ0-02viHm-","executionInfo":{"status":"ok","timestamp":1743023161661,"user_tz":240,"elapsed":35,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"b00726c4-3cec-46b3-9819-267f91a3d5b1"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["5\n"]}]},{"cell_type":"code","source":["# url = \"/content/sample_data/dataset_30312_1.txt\"\n","# with open(url, \"r\") as file:\n","#   my_file = file.readlines()\n","\n","#   new_file = [line.replace(\"\\n\",\"\") for line in my_file]"],"metadata":{"id":"S2HHo5cuikp_"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["pattern = \"GTAGTAA\"\n","DNA = \"CGTTCCGACGTCCCACAAATCGAAGCACGAAGGCGCTCCAACCCAAAGTATCCGCTACCAATAACGGATGGACATAGATTTGAAAACGATTCTCTAGTC ATTAGGCAATTGTCTGCAAAATGTTAACTGTTGAGAAATCTGATTGATTATTCAAGTCAGCGGATTGAAGTAGATCGCGTGGCCTCGACGCGCAGTGGT CTCGAAAGCCTGTCACATCAAGGGGAGCGAGATCAAAGTGACGACTTGCACACCCCTCCGCGGATACTGACACACGATCGCTTGTGATGACGCGGTCGA ATATATATAAACCAATTTATTTAAGTATTACGTACTCCATCTGGATTAAAATTCTGATTGTTAGAACAGGTGCTAGTGCCGCATGTGGAGCGCGAGCAA ACCGGTTCTAGCTTTTATTGGTATGTTGGGGTTAGTAAGCCTGTAACACTATACGTATGGCTTGAGCTTAAATGTTGTGCATTTTCGGTCGGGAGCGCC CGCCCATGTATGTCGGTGCTTAGCCAAACGTTTGTATACGGATTAGTGTGTGTAATAAATCATCGAATCGCCGCAATACAAAGGACCCCATGGGAGCTA TCTTCTGCGGCTTTTTCGGGACTTACGAGGGAGGGGGACCCGTTATGTCTAGAAATGAAACGGTAATAAACAGCAGGGCTGGACAAGTGACGGTTCTTG CTTATTTCTTCAATAAAGATGGACTTTTCCCCCTGAGCTCGGGGCCATGGGTCAGGCACTATCGACTTCAACTCCCAATGACCGGTCTCGTCGCCGTTG GTTCTTTCACCAAGTTCACCCAAAGTAAGATGCGCTAAAAATAAAAATCCACTAGTTCAAGGACTTAACTATCTGATCGCCACGTTCATGTGCGAGCAT CGGGTCCGAAATACTTCGAGAAGGTAGCACCAATTGATATCCATAACTATTAGCCCAGGTTCCGTAATAGAGTCGATGGGAATCACAGATATGGGTTGT TTGATATAATTCAGTCTAGCACGTTGCTGCCCTTGATTTTAGGATCATCACATGACGCCCCGCGCGAATCTAGTGAATGGCTTTCCCAAGGTAGATTCT CAAGTGATAGAAGAGGAGTCCCCTTACTATGGGTGGAATGTTTGTGTTTCGACACAAAGGGAGTAGGCAAGCGGCCCAAAAATTACTAGGGTAGTGGAT ACGTGAACCCTGAGGAACCACGGCTAAGGCACGGCTCATGGAGAACTCGCCTTCCAGATCCAGTGCCCTAGAAGTTGGATTCCTTTGATTGTATAGCTA AGTATATAGTAAGTACACGTCTTACTAGAATCCTCATAGTAGCGGAAGTCCTCTCCGAAGTACAACCCCACTAAGTACTTCGCTCTGTAGAGCATTTAC AGAAATAAACGCTAAAATTCTATTGTCGTATTCACGATTATACTTCCCGAATGCACCACAAATGATTACCGGTTGGTGTTTCCTCATTCACGCAGCACA GCATCCGGCGAGACCCAATCCCGAGTCCGTAAATTACGCGTGTGTAACAAAAAGCTCGCTTGTTATAGTGTAGTGGCGGTTGGTTGCTCCATGGCTGTG TTTTATGCAGCGAGGCTGCTATACTCGTTATCAGAAGTCGATGGTGGCGGGGGCCCGCCTACCCTTTTGTGGACCCAGTTAGCTTCCAAAGGGGAGGTT GCTAACCACCATGATTTAGTGCTAACGGAACAGCATTCCTTCCCAGCCGAGACTAGCGACTGTAGTGATGCGACCGTACGATGTCAGCGATTGCTCAAT GGTAATTTGGGGAGGCCGGCTAAGAGGGAATATAACTGCTGATTTATAAGGGGCCCGGAAAAGTACCTCGGAATTTTACAAACGAATGTGATTGGAGGA CAGGCGTTGAGCAGTCCGCATGATTGGTAGAGCGTCCCTCATGAGCTAATAAATCGAGGAATATCAAGTGAGGGAACGGTTCGTCTGCCGGCTGCTGAT TGGCATTAACACGTTAACGTACCGGTATCTAGCCACCCCCGTCTTGCAGGGCGCGGCTCTATTTCGGGCAGCTGATGCTACCGCCTGATAGAATAAGGA TGGTAATCCGACCCTTGACATTGGGTTAACAACTGCAACGCATATGCGACCGGCGGTAAACAAGGGCACGGCATAATATCCAAAGACACTCATGATATA GAAGCCACGATCATAATACCGTCACACCATGTAAGGGTACTATGGTGTAAGGGCATGACGCATGATACGTGGGATCGGCGTTACACAAAACGGAGGAAT ACACTTATTTTCTTCGTAACCTATAAGTTTACACTCCTATAATGCTCAGCTATTCGCTCTTTCTGCCTCGTGGAGTTAGGACAGTCGAAGGTCTCGTGC CTTCTGTTTTAAGCCATGGCTACCTCCTGCAGCACGAAATTATCCGTCTAGCCATCCCTATTCAGGGGGGCGTGGCCGTACACTAACTAAGCCCTCCGA TGTACACCCATTATATGTCGGATTCTAAGGCAGTAGATTGATTTAGCTGGTCTACGTAAGCAACGATCTCTTGCTAGAGTTTTCCAATGTAGCATGTGT CAATATACCTGTGCGATCCCATAGTGTTAACCGTCGCAAGAGTGTTCTATACTGTCAATCAGTAGACGGCCGCGCTGTAGAGAACACCCATACCCCAGG CCATGGTAACCTGGTAGTGGTTCAACATCCCTATGACGCCCAGCTGAATGGGTTTGTCATTCAAACAATCTCGTAGTCTCTTTCAGCTTGGGCCCCTGC GGCGCGAGCGACGCCTTAAGTTAGGTGGTCAATCGACTGTCCGCGTGAACTTACAGCTGCTTCGCCGGTTGCACTATCAAAGTACCTACCGAACCTGTA CAATTACGCTTTGGGACACTATTTAATCTATCAGCTTAATTAAGATCACTAGATTTCAACCTAGGGGCTGGATGTTAGATTGATAATGAACCCCTAGCC GGAAGAGTGTCTTTATCCTCAACGTGAGGTGCAACCTTCAAGTTGATAGGCGCAGGATTGATCGACTCTGCAAGGATTCACATGAGATGTGGTTCGTTC\"\n","\n","answer = DistanceBetweenPatternAndStrings(pattern,DNA)\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ljkohY_fjC6A","executionInfo":{"status":"ok","timestamp":1743023161783,"user_tz":240,"elapsed":46,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"04788d50-a539-4748-c325-5d8f4bf895ab"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["65\n"]}]},{"cell_type":"markdown","source":["Now that we have a function to compute the distance between Pattern and all pattern' in each string in DNA, we can start looking for the median string, i.e., the string that minimizes the distance between Pattern and all pattern'. <br><br>\n","However, we are not given the Pattern that is capable of this, therefore we have to search for the Pattern that minimzes d(Pattern,DNA)."],"metadata":{"id":"PhCx1zj2kzUr"}},{"cell_type":"code","source":["my_string = \"HELLO\"\n","\n","print(my_string[:0]+\"B\"+my_string[0+1:])\n","\n","for i in range(len(my_string)+1):\n","  print(my_string[:i])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"KmLfqJ8yrKIC","executionInfo":{"status":"ok","timestamp":1743023161844,"user_tz":240,"elapsed":56,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"d9aac170-ddea-49a8-f669-ff49cefd79db"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["BELLO\n","\n","H\n","HE\n","HEL\n","HELL\n","HELLO\n"]}]},{"cell_type":"code","source":["# AllStrings creates all possible k-mers of length k.\n","# INPTUT: integer k for the length of kmers\n","# OUTPUT: A list of all possible kmers of length k\n","\n","def AllStrings1(k: int) -> list[str]:\n","  patterns = set()\n","  nucleotides = [\"A\",\"C\",\"T\",\"G\"]\n","  pattern = \"\"\n","  loop = k\n","  # establishing first_pattern\n","  while loop > 0:\n","    pattern = pattern + \"A\"\n","    loop = loop - 1\n","\n","  patterns.add(pattern)\n","\n","  print(pattern)\n","  neighborhood = Neighbors(pattern, k)\n","\n","\n","  # now with our first string, \"A...A\" of length k, we remove\n","\n","  # for i in range(len(pattern)):\n","  #   for replacement in nucleotides:\n","  #     print(replacement)\n","  #     new_pattern = pattern[:i] + replacement + pattern[i+1:]\n","  #     patterns.add(new_pattern)\n","\n","\n","  return neighborhood"],"metadata":{"id":"xmjv-kjHkEzR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Improved version of AllStrings that essentially rifts off of out Neighborhood function\n","# However, there is no limiting factor in HammingDistance we are looking for, so we will not have the if statement for HD\n","\n","def AllStrings(k: int) -> list[str]:\n","  # our \"ingredients\" for different kmers\n","  nucleotides = [\"A\",\"T\",\"C\",\"G\"]\n","\n","  # our k is one, than all kmers are just our nucleotides,\n","  # this our base case, once this is returned, the actual recursion begins\n","  if k == 1:\n","    return nucleotides\n","\n","  # now we set up for our neighborhood\n","  neighborhood = set() # creating empty set\n","  SuffixPattern = pattern[1:] # Dropping the first symbol in pattern to get the SuffixPattern\n","\n","  # INCREMENT DOWN K, TO MEAT OUR BASE CASE\n","  k -= 1\n","  SuffixNeighbors = AllStrings(k) # Recursion, will keep calling function until base case (len(pattern) == 1)\n","\n","  for text in SuffixNeighbors:\n","    for nucleotide in nucleotides:\n","      # we now add a nucleotide as each possible neigbor to pattern\n","      neighbor = nucleotide + text\n","      neighborhood.add(neighbor)\n","\n","  return neighborhood"],"metadata":{"id":"L2HiRy9DeGgu"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(sorted(AllStrings(2)))\n","\n","# NOT WORKING AS OF 3/10/25"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"hjFILNSUv23b","executionInfo":{"status":"ok","timestamp":1743023161924,"user_tz":240,"elapsed":6,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"b674a668-72e4-415d-b2bd-b42fa8e066d9"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['AA', 'AC', 'AG', 'AT', 'CA', 'CC', 'CG', 'CT', 'GA', 'GC', 'GG', 'GT', 'TA', 'TC', 'TG', 'TT']\n"]}]},{"cell_type":"markdown","source":["Now that we have a function to compute all possible kmers of length k, and a function to minimize the distance (HammingDistance) between one of these found patterns and a set of sequences, we can now look for the MedianString."],"metadata":{"id":"Eze7hXzRgkHq"}},{"cell_type":"code","source":["# MEDIAN STRING\n","# INPUT: space separated string DNA and int k.\n","# OUTPUT: A k-mer Pattern that minimizes d(Pattern, Dna) among all possible choices of k-mers. (If there are multiple such strings Pattern, then you may return any one.)\n","\n","def MedianString(DNA: str, k: int) -> list[str]:\n","  patterns = AllStrings(k)\n","  medians = []\n","\n","  # base case is the max possible distance, in this case the max distance is equal to k\n","  median_distance = k\n","\n","  for pattern in patterns:\n","    distance = DistanceBetweenPatternAndStrings(pattern, DNA)\n","    if distance < median_distance:\n","      median_distance = distance\n","      median = pattern\n","      medians.append(median)\n","\n","  return medians\n","\n","\n"],"metadata":{"id":"U5zV__Zkv5GP"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["dna = \"CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG\"\n","k = 7\n","\n","answer = MedianString(dna,k)\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BAbIUsTMDZ-S","executionInfo":{"status":"ok","timestamp":1743036335964,"user_tz":240,"elapsed":2684,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"e2939b00-314c-4661-aca5-95265d2ca409"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["AATCCTA\n"]}]},{"cell_type":"code","source":["# TESTING MEDIAN STRING\n","k = 3\n","DNA = \"AAATTGACGCAT GACGACCACGTT CGTCAGCGCCTG GCTGAGCACCGG AGTTCGGGACAG\"\n","\n","answer = MedianString(DNA,k)\n","print(answer)"],"metadata":{"id":"rvWRQIgywC6B","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1743023161968,"user_tz":240,"elapsed":2,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"f47870d6-b3df-4c4b-b542-55adc273930e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["GAC\n"]}]},{"cell_type":"code","source":["# url = \"/content/sample_data/dataset_30304_9.txt\"\n","\n","# with open(url, \"r\") as file:\n","#   my_file = file.readlines()\n","\n","# k = int(my_file[0])\n","# DNA = my_file[1]\n","\n","# answer = MedianString(DNA,k)\n","# print(answer)\n","\n","# # IT WORKED!"],"metadata":{"id":"dSAERYS4pIj6"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Quotes from \"Bioinformatics Algorithms: An Active Learning Approach\" by Phillip Compeau and Pavel A. Pevzner:\n","\n","> \"To see why we reformulated the Motif Finding Problem as the equivalent Median String Problem, consider the runtime of MedianString and BruteForceMotifSearch. The former algorithm computes d(Pattern, Dna) for each of the 4k k-mers Pattern. Each computation of d(Pattern, Dna) requires a single pass over each string in Dna, which requires approximately k · n · t operations for t strings of length n in Dna. Therefore, MedianString has a running time of O(4k · n · k · t), which in practice compares favorably with the O(nt · k · t) running time of BruteForceMotifSearch because the length of a motif (k) typically does not exceed 20 nucleotides, whereas t is measured in the thousands.\"\n","\n","> \"The Median String Problem teaches an important lesson, which is that sometimes rethinking how a problem is formulated can lead to dramatic improvements in the runtime required to solve it. In this case, our simple observation that Score(Motifs) could just as easily be computed row-by-row as column-by-column produced the faster MedianString algorithm.\"\n","\n","> \"Of course, the ultimate test of a bioinformatics algorithm is how it performs in practice. Unfortunately, since MedianString has to consider 4k k-mers, it becomes too slow for the Subtle Motif Problem, for which k = 15. We will run MedianString with k = 13 in the hope that it will capture a substring of the correct 15-mer motif. The algorithm still requires half a day to run on our computer and returns the median string AAAAAtAGaGGGG (with distance 29). This 13-mer is not a substring of the implanted pattern AAAAAAAAGGGGGGG, but it does come close.\""],"metadata":{"id":"GWYZanmVrISz"}},{"cell_type":"markdown","source":["#### GREEDY MOTIF SEARCH 1.5"],"metadata":{"id":"INwHYJm9txu8"}},{"cell_type":"markdown","source":["A \"Greedy Algorithm\" will select the \"best\" looking answer for each *individual* step within an algorithm. However, this scope can be too narrow, leading to inaccurate solutions in the end. Another approach would be a \"Heuristic Algorithm\", which would be less accurate but perform much faster to find an approximate solution. Both have their place when it comes to using or creating an algorithm, but the use of each is contextual."],"metadata":{"id":"edtXQNwsHVr3"}},{"cell_type":"markdown","source":["We will be using a \"Greedy Approach\" for finding a possible Motif by using the probabilty matrix. The higher the probability of a kmer, it is more liekly to be similar to the consensus string."],"metadata":{"id":"7NlJ4z-IH-QC"}},{"cell_type":"code","source":["# # --- OLD VERSION ---\n","\n","# # ProfileMostProbableKmer -> We will find a Profile-Most probable kmer of a string\n","# # INPUT: a string TEXT, integer k, and a 4xk profile matrix (row x column)\n","# # OUTPUT: A Profile-most probable kmer in TEXT.\n","\n","# # we are basing out rows from previous examples, order is as follows:\n","# # 0 -> A, 1 -> C, 2 -> G, 3 -> T\n","# def ProfileMostProbableKmer(text: str, k: int, profiles) -> str:\n","#   mostProbableKmer = text[:k]\n","#   kmers = set()\n","#   prob_table = {}\n","#   # lets list our kmers\n","#   for i in range(len(text)-k+1):\n","#     kmer = text[i:i+k]\n","#     kmers.add(kmer)\n","\n","#   # for each kmer, we want to compute its probability based on the given probabilty matrix\n","#   for kmer in kmers:\n","#     # setting the base probability as 1, but this will change accordingly when multiplied by the actual probability.\n","#     prob_table[kmer] = 1\n","\n","#     # for each nucleotide in kmer, we check which it is and then multiply the appropriate probability based on its position in the string.\n","#     for nucleotide in range(k):\n","#       if kmer[nucleotide] == \"A\":\n","#         prob_table[kmer] = (prob_table[kmer] * profiles[0][nucleotide])\n","\n","#       if kmer[nucleotide] == \"C\":\n","#         prob_table[kmer] = (prob_table[kmer] * profiles[1][nucleotide])\n","\n","#       if kmer[nucleotide] == \"G\":\n","#         prob_table[kmer] = (prob_table[kmer] * profiles[2][nucleotide])\n","\n","#       if kmer[nucleotide] == \"T\":\n","#         prob_table[kmer] = (prob_table[kmer] * profiles[3][nucleotide])\n","\n","#   # now we look for the max probability value in our table\n","#   max_val = max(prob_table.values())\n","\n","#   # now we look for the kmer that has that max value\n","#   for kmer in prob_table:\n","#     if prob_table[kmer] == max_val:\n","#       mostProbableKmer = kmer\n","\n","#   return mostProbableKmer"],"metadata":{"id":"YGC67WkqphVS"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# -- UPDATED VERSION -- 03/25/25\n","# ProfileMostProbableKmer -> We will find a Profile-Most probable kmer of a string\n","# INPUT: a string TEXT, integer k, and a 4xk profile matrix (row x column)\n","# OUTPUT: A Profile-most probable kmer in TEXT.\n","\n","# we are basing out rows from previous examples, order is as follows:\n","# 0 -> A, 1 -> C, 2 -> G, 3 -> T\n","def ProfileMostProbableKmer(text: str, k: int, profiles) -> str:\n","  mostProbableKmer = text[:k]\n","  kmers = set()\n","  prob_table = {}\n","  # lets list our kmers\n","  for i in range(len(text)-k+1):\n","    kmer = text[i:i+k]\n","    kmers.add(kmer)\n","\n","  # for each kmer, we want to compute its probability based on the given probabilty matrix\n","  for kmer in kmers:\n","    # setting the base probability as 1, but this will change accordingly when multiplied by the actual probability.\n","    prob_table[kmer] = 1\n","\n","    # for each nucleotide in kmer, we check which it is and then multiply the appropriate probability based on its position in the string.\n","    for column in range(k):\n","      # UPDATE\n","      # using the .get() method allows us to only obtain the probability in profile correspondin to the nucleotide and position of kmer[column]\n","      prob_table[kmer] = (prob_table[kmer] * profiles[column].get(kmer[column],0))\n","\n","  # now we look for the max probability value in our table\n","  max_val = max(prob_table.values())\n","\n","  # now we look for the kmer that has that max value\n","  for kmer in prob_table:\n","    if max_val == 0:\n","      return mostProbableKmer\n","\n","    if prob_table[kmer] == max_val:\n","      mostProbableKmer = kmer\n","\n","  return mostProbableKmer"],"metadata":{"id":"tdR-6pVxIMdB"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# testing the different methods for grabbing indices and values from maps/dicts\n","\n","profile = [[1,2,3,4],[5,6,7,8]]\n","\n","for row, index_row in enumerate(profile):\n","  print(\"row:\",row,\"index row:\",index_row)\n","\n","\n","my_dict = {\"A\":0.4,\"C\":0.6}\n","print(my_dict[\"A\"])\n","\n","for letter in my_dict:\n","  print(\"letter:\",letter, \"value:\",my_dict[letter])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0OmCbGiPbt73","executionInfo":{"status":"ok","timestamp":1743023162047,"user_tz":240,"elapsed":3,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"173d8504-351a-4d0c-a0ad-b96e124b186c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["row: 0 index row: [1, 2, 3, 4]\n","row: 1 index row: [5, 6, 7, 8]\n","0.4\n","letter: A value: 0.4\n","letter: C value: 0.6\n"]}]},{"cell_type":"code","source":["# tesing MostProbableKmer function\n","text = \"ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT\"\n","k = 5\n","# prob_matrix_1 = [[0.2, 0.2, 0.3, 0.2, 0.3],\n","#                [0.4, 0.3, 0.1, 0.5, 0.1],\n","#                [0.3, 0.3, 0.5, 0.2, 0.4],\n","#                [0.1, 0.2, 0.1, 0.1, 0.2,]]\n","# changed the desired profile format to be a dictionary, where the keys are the\n","# column indices, and the values are the probabilities for each nucleotide in that column\n","prob_matrix_2 = {0:{\"A\":0.2,\"C\":0.4,\"G\":0.3,\"T\":0.1},\n","                1:{\"A\":0.2,\"C\":0.3,\"G\":0.3,\"T\":0.2},\n","                2:{\"A\":0.3,\"C\":0.1,\"G\":0.5,\"T\":0.1},\n","                3:{\"A\":0.2,\"C\":0.5,\"G\":0.2,\"T\":0.1},\n","                4:{\"A\":0.3,\"C\":0.1,\"G\":0.4,\"T\":0.2}}\n","\n","answer = ProfileMostProbableKmer(text,k,prob_matrix_2)\n","print(answer)"],"metadata":{"id":"9mVGEzaKb4oy","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1743023162084,"user_tz":240,"elapsed":37,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"32f6a0a2-2c6c-4081-ed75-aa07e518a7b1"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["CCGAG\n"]}]},{"cell_type":"code","source":["prob = max(prob_matrix_2[0].values())\n","nucleotide = max(prob_matrix_2[0], key=prob_matrix_2[0].get)\n","print(prob)\n","print(nucleotide)\n","print(prob_matrix_2[0].get)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5oWqBzGSLQh9","executionInfo":{"status":"ok","timestamp":1743023162160,"user_tz":240,"elapsed":40,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"e3d7f32b-5b5a-4f7e-d806-46c987ec5e08"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["0.4\n","C\n","<built-in method get of dict object at 0x7ad019c0aa80>\n"]}]},{"cell_type":"code","source":["# url = \"/content/sample_data/dataset_30305_3.txt\"\n","\n","# with open(url, \"r\") as file:\n","#   my_file = file.readlines()\n","\n","#   new_files = [line.replace(\"\\n\",\"\") for line in my_file]\n","\n","#   # k = int(new_files[0])\n","#   # text = new_files[1]\n","\n","# text = new_files[0]\n","# k = int(new_files[1])\n","\n","# prof_matrix = []\n","# for line in new_files[2:]:\n","#   prof_matrix.append([float(i) for i in line.split()])\n","# answer = ProfileMostProbableKmer(text,k,prof_matrix)\n","# print(answer)"],"metadata":{"id":"ve1GCstIqqw-"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["`\"Our proposed greedy motif search algorithm, GreedyMotifSearch, starts by forming a motif matrix from arbitrarily selected k-mers in each string from Dna (which in our specific implementation is the first k-mer in each string). It then attempts to improve this initial motif matrix by trying each of the k-mers in Dna1 as the first motif. For a given choice of k-mer Motif1 in Dna1, it builds a profile matrix Profile for this lone k-mer, and sets Motif2 equal to the Profile-most probable k-mer in Dna2. It then iterates by updating Profile as the profile matrix formed from Motif1 and Motif2, and sets Motif3 equal to the Profile-most probable k-mer in Dna3. In general, after finding i − 1 k-mers Motifs in the first i − 1 strings of Dna, GreedyMotifSearch constructs Profile(Motifs) and selects the Profile-most probable k-mer from Dnai based on this profile matrix. After obtaining a k-mer from each string to obtain a collection Motifs, GreedyMotifSearch tests to see whether Motifs outscores the current best scoring collection of motifs and then moves Motif1 one symbol over in Dna1, beginning the entire process of generating Motifs again.\"`"],"metadata":{"id":"v_XtecL5xDC5"}},{"cell_type":"markdown","source":["Now to use our greedy motif search algorithm, we will start by collecting \"randomly\" selected motifs from each DNA string; for our implementation it will be the first kmer within each DNA string. <br>\n","\n","The function will then attempt to improve the kmer selection by building a profile matrix Profile for this kmer, and Motif2 will be the Profile-most probable kmer in DNA2, and Motif3 will be the DNA most probable kmer in DNA3 and so on and selects the motif that maximizes the Profile within DNAi. <br>\n","\n","Essentially, our GreedyMotifSearch tests the Motifs against the current best scoring collection of motifs, moved Motif1 one nucleotide down and starts the whole process all over again, generating another batch of motifs to compare against \"BestMotif\" (that was selected from the previous batch).\n"],"metadata":{"id":"WoM7bJOfxLnh"}},{"cell_type":"markdown","source":["\n","\n","```\n","# PseudoCode for GreedyMotifSearch\n","def GreedyMotifSearch(DNA, k , t):\n","  BestMotifs <- Motif Matrix formed by first kmer in each string\n","  FOR each kmer motif in the first string of DNA\n","    Motif_1 <- Motif\n","    FOR i = 2 to t:\n","      form Profile from Motif_1...Motif_t-1\n","      Motif_i <- Profile-Most probable kmer from ith string DNA\n","    Motifs <- (Motifs_i...Motifs_t)\n","    IF score(Motifs) < score(BestMotifs)\n","      BestMotifs <- Motifs\n","    \n","  return BestMotifs\n","\n","```\n"],"metadata":{"id":"IVCN-PXcACvF"}},{"cell_type":"code","source":["motif = [\"ACCT\"]\n","display(countMotifPercent(motif))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":86},"id":"1HxHWkCtBb1R","executionInfo":{"status":"ok","timestamp":1743023162280,"user_tz":240,"elapsed":78,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"3caddde0-5bdf-4c7a-84be-b6b79eb4d55f"},"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["{0: {'A': 1.0, 'C': 0.0, 'G': 0.0, 'T': 0.0},\n"," 1: {'A': 0.0, 'C': 1.0, 'G': 0.0, 'T': 0.0},\n"," 2: {'A': 0.0, 'C': 1.0, 'G': 0.0, 'T': 0.0},\n"," 3: {'A': 0.0, 'C': 0.0, 'G': 0.0, 'T': 1.0}}"]},"metadata":{}}]},{"cell_type":"code","source":["# # GreedyMotifSearch\n","# # INPUT: integers k,t, followed by space separated colletion of strings DNA\n","# # OUTPUT: A collection of strings BestMotifs resulting from applying GreedyMotifSearch(Dna, k, t).\n","# #         If at any step you find more than one Profile-most probable k-mer in a given string, use the one occurring first.\n","\n","# def GreedyMotifSearch(DNA: str, k: int, t: int) -> list[str]:\n","#   bestMotifs = []\n","#   dnaList = DNA.split()\n","#   consensusMotifs = []\n","\n","#   # # creating our first set of \"best\" motifs from the first kmers in each DNA string\n","#   # for string in dnaList:\n","#   #   bestMotifs.append(string[:k])\n","\n","#   for i in range(len(dnaList[0])-k+1):\n","#     motif = dnaList[0][i:i+k]\n","#     bestMotifs = [motif]\n","\n","#     for j in range(1,len(dnaList)):\n","#       profile = countMotifPercent(bestMotifs)\n","#       motif = ProfileMostProbableKmer(dnaList[j],k,profile)\n","#       bestMotifs.append(motif)\n","#\n","\n","\n","#       if j == len(dnaList)-1:\n","#         consensus_motif = \"\"\n","#         for column in range(k):\n","#           # append the the nucleotide with the highest probability in each column\n","#           consensus = max(profile[column], key=profile[column].get)\n","#           consensus_motif = consensus_motif[:column] + consensus + consensus_motif[column+1:]\n","#           if column == k-1:\n","#             consensusMotifs.append(consensus_motif)\n","\n","            # NEEDED TO COMPUTE THE MOTIF SCORE HERE, THEN IF IT IS LOWER, HAVE THE CURRENT SET OF MOTIFS BE THE \"BEST MOTIFS\"\n","\n","#   return consensusMotifs\n","\n","#   # A C G T\n","\n","#   return bestMotifs\n"],"metadata":{"id":"lz9Z5xGPwmFw"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# --- UPDATED VERSION ---\n","# 03/26/25\n","\n","def GreedyMotifSearch(DNA: str, k: int, t: int) -> list[str]:\n","    dnaList = DNA.split()\n","    bestMotifs = [string[:k] for string in dnaList]  # Initialize with first k-mers\n","    bestScore = float('inf')  # Track the best score (lower is better)\n","\n","    # Iterate over all possible k-mers in the first DNA string\n","    for i in range(len(dnaList[0]) - k + 1):\n","        motifs = [dnaList[0][i:i+k]]  # Start with the current k-mer from the first sequence\n","\n","        # Build the profile and find motifs in the remaining sequences\n","        for j in range(1, t):\n","            profile = countMotifPercent(motifs)\n","            nextMotif = ProfileMostProbableKmer(dnaList[j], k, profile)\n","            motifs.append(nextMotif)\n","\n","        # Compute the score of the current motif set\n","        currentScore = computeMotifScore(motifs)\n","\n","        # Update bestMotifs if the current set is better\n","        if currentScore < bestScore:\n","            bestScore = currentScore\n","            bestMotifs = motifs.copy()\n","\n","    return bestMotifs\n","\n","\n","def computeMotifScore(motifs: list[str]) -> int:\n","    \"\"\"Compute the total number of mismatches against the consensus motif.\"\"\"\n","    if not motifs:\n","        return 0\n","\n","    consensus = []\n","    k = len(motifs[0])\n","\n","    # Build the consensus string\n","    for i in range(k):\n","        column = [motif[i] for motif in motifs]\n","        mostCommon = max(set(column), key=column.count)\n","        consensus.append(mostCommon)\n","    consensusMotif = ''.join(consensus)\n","    print(consensusMotif)\n","\n","    # Calculate total mismatches\n","    score = 0\n","    # for motif in motifs:\n","    #     score += sum(1 for a, b in zip(motif, consensusMotif) if a != b)\n","    for motif in motifs:\n","      score += HammingDistance(motif,consensusMotif)\n","\n","    return score"],"metadata":{"id":"sUpOL4kSRbI6"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["dna = \"GGCGTTCAGGCA AAGAATCAGTCA CAAGGAGTTCGC CACGTCAATCAC CAATAATATTCG\"\n","k = 3\n","t = 5\n","\n","answer = GreedyMotifSearch(dna,k,t)\n","\n","print(answer)\n","# getting too many nucleotides in the result\n","  # expecting 5, but got 4"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"k4nQ_MhUF8W4","executionInfo":{"status":"ok","timestamp":1743024765640,"user_tz":240,"elapsed":11,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"26b4d329-44c0-4ea2-9f97-ab9735b7252a"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['CAG', 'CAG', 'CAA', 'CAA', 'CAA']\n"]}]},{"cell_type":"code","source":["# url = \"/content/sample_data/dataset_30305_5.txt\"\n","# with open (url, \"r\") as file:\n","#   my_file = file.readlines()\n","\n","#   new_files = [line.replace(\"\\n\",\"\") for line in my_file]"],"metadata":{"id":"szfM24LrGnkt"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["my_file = ['12 25',\n"," 'ATGGGCAGCATGGAGTACACCTCACGAATGTTGGGTACGGGCTAGGGGAATGAACCTTTGGTCCGAGAGTTTGCCTACACCACAGTTTAACAGACTTAACTACCACGCTCGTACATAGAGTGTAGGCAGCTGGCTATGCTACGAACCGTAAGGTAG GGAAAGACACAATATCGTTTGGGTGTGCAAATTGTGTGGTGCTAGCATTGAAATGGAGCAGTACCGTGTACCCGAGGGGGTGTCGTTCAAATCCAGGACCCGCGAGGCGACTATCCCACACGGCGTATGAGAACTTTGGACAGCTAGGGATTGCCG GTGTTCACTTGGGAATAGTTTATCGGTATAAGACCATGTCTCGCAAACTGGTGTACAGCAGTCAGCAGACGGGGCTTGCCCACATCTACGTACTCCCTGTCTGTGGCTAGTGTAATCACCCGTAGCCGCGATCGATTATGGCGACAATGTTATGGA GGGCCGTTCTTTTCTCTCGCGTAAAACTGAGCGTGCGAAATGTTCCGGCAAAAGCTCCAGTTATCTGGGATCGACTGCCCCACAGGAGGGATTATCCAACGACTACATACGACAGGAGCAGAATAACGTGAGCCGAACCAGTACATGACGGCATCA CGGGATGCGGTATTTTACCGGTCAACACGTTTTCAGAATGCCTTCCAAAATCACCAAAGACAGCTTTGAAATCCACCCTCGTCGTGCTGGTCCGCGATGCGGGACTTTCATTTCGACCACTCCAATAAGCATTTCAGGGGCAGGGACTGGTCCACA GGCACGAAGTCTCAACTCATTACACTTAGGCCTAACCGACTATGGGACGACTGGACCACATACCGAGGGCTGTTAGACCATGTCGCGGAAACTCCTCGAGAATATTGTAGATGCTAATTCGCTCACGCTGCCGTGGTGAGTGGTTGATCCAGCGTT TACCACAGGGGTCGTGTTCGTGGCTGCGGAATGATTTAACAGGCTTGGCGTGGGTGCTAGGGCACTAAGGGCCTAAAAAAGTTCACTAACCGGCTGATAGGGGCGGTTGTTATGGCTTTGGTCTGGGCCACAATTCATGAGAAGTCGCATAGTAAC TTATACCTGACTTTGCTTGAATTTGTGATAGTAGAAGTCTCGCCCACATGTGGACATGTGGGTATTTCGAATAGGAACTGGGGGGTGTATTATCATGGAATCAATTAACATCGGTCAAAGGGGCTCACGCGAATACGAGCCGGTAGCATGCGTTGA CTAAGTATATTCGACTACACCACACTTTCTGCATACAGTCAACCGAGACCCGAACTTCACTATTGTAAGTAACGTTTGAGTTACGTGTAAAAGGTTGGACGCAAATCAGAGTGTGCGTGCCTCGTCACAGAATACCCGGACATGGCAAAATGTGAT GACTGAGCCACAAACATCTCACGTGCTTCTCCTACTAACCCACTGTGGCCTAATCCGAAGTCAGAATCTCTCGGTCCCTGAGATTTCGTGACAACGGTCTCGGCGTATTATTAGCAAAAACGGCGGCGTATAGAGTTATGACTGAACAGAAACATC TGAAAAGTACACAATCGATCGGTGGTAGTGATAGACAAAACTTACCACTCGCAGAGGGCATCCAAGAAAGCTGCCCAGAGAGCAATCAAAACACCATCAGTACGCTCTCCAAGAAGATACAGAAGTCCCACTGACTTGGCCACAAGCACTGTATGT TATCTAATGCTCTGCATTGAACGTTACCCAGTGTCGATATGAACAAGGGCCTCGTCCACACCGCAGAGCTGCCAACCTGAGAGCGGTAGGCAGTCCCTTGGACGGTGCAACTATGGCGCTGGGGTAACTAAGAGAAGGTCTATCGTGAAGCGCTAT AGTTGAATATTTTACTGAATCAATCGATAACTTCTATTCCCTGCCCTAATTGCCCCACGTGACTGTACCACACACGAGCTATCATACGGGTGTGCATGTTCCCTGACCGAGCAAGCGTTCGACAGCAAACCCCCCGGTGATAACATACCCCCTGCA GTAGGCCTGCAGGCCTTGGCCACAGGTGGCCGAGTGCAGCTCATGAGTAGCGCAACTCAGCGACGCGTCGTTATGTGAAGACACAAACCCCCAAACTTCCGCAACAACTCCAGCCTGGTCGCGGCACTACTGACATAAGACGGGTTCGGCTGGAAG AAGCCGATGTAAAAGCATATGGCCCTATGGTACACCAGAAGGTAACTAGTCTCCCGCATAACAGCGGACATCGGTGAACCACAAGAAGCATCGGCCTGCTCCCTAGCCTCCAAATTATCGGAGATCTATCGCGTCTAGCCCACAACACGGGAGTGT GTATCATGTTAAAAATGCGTGTCGTCACCCAACACCCCCACGGGCACCGTCTCAGCCACAAAAGGGTAGCGATTGTTGATCGGTGCTTAATGCGGAGTATTCCTGCTAACTTGTCCGAGGTAAAGGACCCAATGTGCTCTAAATCTACGTTGGATA TATACCGGTCGAGCTTACTGAATCCCAATGTTCGCTCGCGGGACAGGAATAGTGGAACGCTTGGCACTTGCGAATACTGTGCGGGTCTGACCCACAACGGCGAACAATTCGTGTCCGTTGATCTACGGTATCCGTTATAATCCCACGGTGGGCAAT CTGATTACTTACCGTGGCAGGTTTTGTGGCGTAGGCCTTTCATCGTACTTCTGCCGTTTTGACAAATATGATTGTAAAATCAAACTCGTGACATTCGCAGCATTAGTTTGGCGCGAATTCCTCGTTCACTGCGCCTTTCCCACAGAGGCGGGCAAT TATTATAGTCGGGCGCTTCGAATGAAAATATTGGTAAAAAGAGACAGCAGAGTTAAAATTTACCGCATTCCATATCCACTCTTCGGACATCCTCTACGGATATAATTCACTGAACAAGTTGCCTTAGCCACATCTATTTGCGGCTCTGTGGAAAAT AAAGCCGTGGATCCTGTGGGCGATAGTAGCCCTTCGTTTAAGGAGCGCAGCGCGAGGTGTCCCCACCCGGATGACTTCTCCACATATAGCCCCCCGCGCAATCACATAACGGAATCCGCGTAATGGTTGAGGTCTGTTTGTTGGGACGGGGGGAAG TGTTCCAGACCGGCTGTGCTAGTAGGCTCATCCACAGGTCCGTGACGCGGGTTCTCGGGTTTGAGTCTCCCCGGTCTTGCTGCGGACAACCCGCGAGTGTGTTGTCATTGCCGATAAAAGCCGCTGTCTCGTGTCACGGTACGTCTGAGCGGTCTA ACGCTGTATTTACAAGCGGCCTCTATCGGGGTGTCTAGGACTAATGTTTAGAGTGGTATGAACCTCATTCACGCAGCATACGTTGGAATTGTCCTGAGAACGCTCTACCGATCTCTACATACAAGCCCTGCAGCCAGCAGGGGTGCCTACACCACA ATGCTCGCCCTTAACCGGACCCATCGCATAAGTCCTATTGGTTAGGCTCGACTCTCGTTAGAGATACCTCACGACGCCTACCACTACCAGTACCTAACAGGTGACACAGCCGCTTACAAGGTCTGTCCCACAGGCACGGGGTGTCCGCGCGATTAT GCGGGTTTTGTGCGAAATTGTATAAACTAGCCTCAGTCAACGAACTGCACACAGAGAAGTGGATACGCCCCGACATCAACTCAGGCAACCTACAGGACAACGTCGCCGGGCTACTCCACACCCAGAATGCAGAAGTATGATGTATGTACTGTATAT TAGTCCGGCGGTACCCGCGTAAAGGGCGTTGTCATTTCTACGCACTCGATATGTAAGTACGGGACTTCGGCCCGTAGGCCCGTGGTCTGGTCCACATAAGAGCCATGGTCACACGCTCCAAATCCATGGGGCAGATTGAGAAAGGTAAGAGGTGTT']"],"metadata":{"id":"bB5xxDSZoWAB"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# new_files[0] = new_files[0].split()\n","# k = int(new_files[0][0])\n","# t = int(new_files[0][1])\n","# dna = new_files[1]\n","\n","# answer = GreedyMotifSearch(dna,k,t)"],"metadata":{"id":"bi8MwjiooRqz"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"WoubDOB2odFa","executionInfo":{"status":"ok","timestamp":1743029133969,"user_tz":240,"elapsed":21,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"1881e859-b3de-4e65-9914-db644b334860"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['TAGGGGAATGAA', 'GGAAAGACACAA', 'GTGTTCACTTGG', 'GGGCCGTTCTTT', 'CGGGATGCGGTA', 'GGGACGACTGGA', 'GGAATGATTTAA', 'GGAATCAATTAA', 'GTAACGTTTGAG', 'GAGATTTCGTGA', 'GAGAGCAATCAA', 'GTGTCGATATGA', 'TGAATCAATCGA', 'GTGAAGACACAA', 'CAAATTATCGGA', 'TAAAGGACCCAA', 'GAGCTTACTGAA', 'CTGCCGTTTTGA', 'TAATTCACTGAA', 'GTAATGGTTGAG', 'GTGTTGTCATTG', 'GAGTGGTATGAA', 'CAGGTGACACAG', 'GAAATTGTATAA', 'GGGCAGATTGAG']\n"]}]},{"cell_type":"code","source":["display(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":451},"id":"MSc3fMjQoeHy","executionInfo":{"status":"ok","timestamp":1743029139239,"user_tz":240,"elapsed":4,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"e77ee5a2-713d-4112-a8bc-e85ceae3a49c"},"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["['TAGGGGAATGAA',\n"," 'GGAAAGACACAA',\n"," 'GTGTTCACTTGG',\n"," 'GGGCCGTTCTTT',\n"," 'CGGGATGCGGTA',\n"," 'GGGACGACTGGA',\n"," 'GGAATGATTTAA',\n"," 'GGAATCAATTAA',\n"," 'GTAACGTTTGAG',\n"," 'GAGATTTCGTGA',\n"," 'GAGAGCAATCAA',\n"," 'GTGTCGATATGA',\n"," 'TGAATCAATCGA',\n"," 'GTGAAGACACAA',\n"," 'CAAATTATCGGA',\n"," 'TAAAGGACCCAA',\n"," 'GAGCTTACTGAA',\n"," 'CTGCCGTTTTGA',\n"," 'TAATTCACTGAA',\n"," 'GTAATGGTTGAG',\n"," 'GTGTTGTCATTG',\n"," 'GAGTGGTATGAA',\n"," 'CAGGTGACACAG',\n"," 'GAAATTGTATAA',\n"," 'GGGCAGATTGAG']"]},"metadata":{}}]},{"cell_type":"code","source":["display(\" \".join(answer))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":70},"id":"rEa6dmF4ofaZ","executionInfo":{"status":"ok","timestamp":1743029200707,"user_tz":240,"elapsed":5,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"b95d15ec-0e48-4a1e-c8f6-2d36a3bae32c"},"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["'TAGGGGAATGAA GGAAAGACACAA GTGTTCACTTGG GGGCCGTTCTTT CGGGATGCGGTA GGGACGACTGGA GGAATGATTTAA GGAATCAATTAA GTAACGTTTGAG GAGATTTCGTGA GAGAGCAATCAA GTGTCGATATGA TGAATCAATCGA GTGAAGACACAA CAAATTATCGGA TAAAGGACCCAA GAGCTTACTGAA CTGCCGTTTTGA TAATTCACTGAA GTAATGGTTGAG GTGTTGTCATTG GAGTGGTATGAA CAGGTGACACAG GAAATTGTATAA GGGCAGATTGAG'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{}}]},{"cell_type":"markdown","source":["#### MOTIF FINDING MEETS OLIVER CROMWELL 1.6"],"metadata":{"id":"MFoFkFTxpQT7"}},{"cell_type":"markdown","source":["We will be taking a look at **Cromwell's Rule**, which advises against using 1 or 0 for prior probabilities unless only using it for logical expressions that can only be true or false.<br>\n","\n","Introducing the use of **pseudocounts**. By adding 1 to the number of occurences of ALL events, we will essentially replace any zero probabilities with a non-zero probability."],"metadata":{"id":"MQlpKHc1p4Gx"}},{"cell_type":"code","source":["# rewriting Greedy Motif Search to include pseudocounts\n","# --- UPDATED VERSION ---\n","# 03/26/25\n","\n","def GreedyMotifSearchWithPseudocounts(DNA: str, k: int, t: int) -> list[str]:\n","    dnaList = DNA.split()\n","    bestMotifs = [string[:k] for string in dnaList]  # Initialize with first k-mers\n","    bestScore = float('inf')  # Track the best score (lower is better)\n","\n","    # Iterate over all possible k-mers in the first DNA string\n","    for i in range(len(dnaList[0]) - k + 1):\n","        motifs = [dnaList[0][i:i+k]]  # Start with the current k-mer from the first sequence\n","\n","        # Build the profile and find motifs in the remaining sequences\n","        for j in range(1, t):\n","            profile = countMotifPercentPseudocounts(motifs)\n","            nextMotif = ProfileMostProbableKmer(dnaList[j], k, profile)\n","            motifs.append(nextMotif)\n","\n","        # Compute the score of the current motif set\n","        currentScore = computeMotifScore(motifs)\n","\n","        # Update bestMotifs if the current set is better\n","        if currentScore < bestScore:\n","            bestScore = currentScore\n","            bestMotifs = motifs.copy()\n","\n","    return bestMotifs\n"],"metadata":{"id":"kF6jseE4paEo"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def countMotifPercentPseudocounts(motifs):\n","    count = {}\n","    columns = []\n","    for i in range(len(motifs[0])):\n","        columns.append([motif[i] for motif in motifs])\n","    for i in range(len(columns)):\n","        # adding a psuedocount (+1) to remove probabilities of zero\n","        count[i] = {'A': (columns[i].count('A'))+1/len(columns[i]), 'C': (columns[i].count('C'))+1/len(columns[i]), 'G': (columns[i].count('G'))+1/len(columns[i]), 'T': (columns[i].count('T'))+1/len(columns[i])}\n","\n","    return count"],"metadata":{"id":"TA3HV_1Eu8ki"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["k = 3\n","t = 5\n","dna = \"GGCGTTCAGGCA AAGAATCAGTCA CAAGGAGTTCGC CACGTCAATCAC CAATAATATTCG\"\n","\n","answer = GreedyMotifSearchWithPseudocounts(dna,k,t)\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kdBPVlKuvuBC","executionInfo":{"status":"ok","timestamp":1743031083631,"user_tz":240,"elapsed":4,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"359f4964-5a18-4a45-e0a2-e6b6269a4bba"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['TTC', 'ATC', 'TTC', 'ATC', 'TTC']\n"]}]},{"cell_type":"code","source":["url = \"/content/sample_data/dataset_30306_9.txt\"\n","\n","with open(url, \"r\") as file:\n","  my_file = file.readlines()\n","\n","  new_files = [line.replace(\"\\n\",\"\") for line in my_file]\n","\n","k = int(new_files[0].split()[0])\n","t = int(new_files[0].split()[1])\n","dna = new_files[1]\n","\n","answer = GreedyMotifSearchWithPseudocounts(dna,k,t)\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"J7_hEl3Qv4UZ","executionInfo":{"status":"ok","timestamp":1743031137943,"user_tz":240,"elapsed":1569,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"6c2aab8e-ab0a-4bf5-e67a-e7cc9e0f1a70"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['AAGACCCGTGCA', 'CAGCCCCCTGCA', 'TAGACCCTTACA', 'AAGACCCTTCCA', 'GAGTCCCATGCA', 'TAGGCCCGTACA', 'TAGCCCCATTCA', 'CAGACCCGTTCA', 'GAGACCCATCCA', 'TAGTCCCATCCA', 'GAGTCCCTTCCA', 'GAGTCCCCTTCA', 'CAGGCCCTTGCA', 'GAGGCCCCTTCA', 'GAGGCCCGTCCA', 'CAGGCCCTTCCA', 'GAGACCCGTTCA', 'GAGCCCCGTACA', 'CAGCCCCGTGCA', 'CAGCCCCCTTCA', 'CAGGCCCCTGCA', 'CAGCCCCATGCA', 'AAGGCCCATCCA', 'TAGCCCCATACA', 'CAGTCCCTTACA']\n"]}]},{"cell_type":"code","source":["display(\" \".join(answer))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":70},"id":"V-HoS4sgwHAA","executionInfo":{"status":"ok","timestamp":1743031152980,"user_tz":240,"elapsed":47,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"f79c5cbd-d843-40a6-9247-43fbf5b5d0aa"},"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["'AAGACCCGTGCA CAGCCCCCTGCA TAGACCCTTACA AAGACCCTTCCA GAGTCCCATGCA TAGGCCCGTACA TAGCCCCATTCA CAGACCCGTTCA GAGACCCATCCA TAGTCCCATCCA GAGTCCCTTCCA GAGTCCCCTTCA CAGGCCCTTGCA GAGGCCCCTTCA GAGGCCCGTCCA CAGGCCCTTCCA GAGACCCGTTCA GAGCCCCGTACA CAGCCCCGTGCA CAGCCCCCTTCA CAGGCCCCTGCA CAGCCCCATGCA AAGGCCCATCCA TAGCCCCATACA CAGTCCCTTACA'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{}}]},{"cell_type":"code","source":["profile = {0: {'A': 0.4, 'C': 0.2, 'G': 0.1, 'T': 0.3},\n","           1: {'A': 0.3, 'C': 0.3, 'G': 0.3, 'T': 0.1},\n","          2: {'A': 0.0, 'C': 0.0, 'G': 1.0, 'T': 0.0},\n","           3: {'A': 0.1, 'C': 0.4, 'G': 0.1, 'T': 0.4},\n","           4: {'A': 0.0, 'C': 0.0, 'G': 0.5, 'T': 0.5},\n","           5: {'A': 0.9, 'C': 0.1, 'G': 0.0, 'T': 0.0}}"],"metadata":{"id":"KUeDy--gwLCY"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["entropy = 0\n","for i in range(len(profile)):\n","    for nucleotide in profile[i]:\n","        if profile[i][nucleotide] != 0:\n","            entropy += profile[i][nucleotide] * math.log2(profile[i][nucleotide])\n","\n","\n","display(-entropy)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"vPDpkROOhV1v","executionInfo":{"status":"ok","timestamp":1743044253718,"user_tz":240,"elapsed":14,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"5477f140-87d8-47de-d012-59e1e3e39628"},"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["6.932824877385981"]},"metadata":{}}]},{"cell_type":"code","source":["mostCommon = max(profile[0], key=profile[0].get)\n","print(mostCommon)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"x-sgEa5vjUSy","executionInfo":{"status":"ok","timestamp":1743044627574,"user_tz":240,"elapsed":4,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"031b0dea-f980-48e8-848d-ae99b3bfa07d"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["A\n"]}]},{"cell_type":"code","source":["consensus = []\n","for i in range(len(profile)):\n","  mostCommon = max(profile[i], key=profile[i].get)\n","  consensus.append(mostCommon)\n","consensusMotif = ''.join(consensus)\n","print(consensusMotif)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"DGr_LKg1iJeO","executionInfo":{"status":"ok","timestamp":1743044683652,"user_tz":240,"elapsed":12,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"a8d7688b-e6ba-472b-df55-fc8b7cb9524c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["AAGCGA\n"]}]},{"cell_type":"code","source":["motif_matrix = \"CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG\"\n","k = 7"],"metadata":{"id":"evWXzfn7jyb8"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["MedianString(motif_matrix,k)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"7UTGeVz7kv2s","executionInfo":{"status":"ok","timestamp":1743045049629,"user_tz":240,"elapsed":3475,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"376c3705-1b4a-4d9c-b954-6fa44ddb32c1"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['AATCCAG', 'CAATCGG', 'GTTTCAT', 'AGTTTCA', 'GTAGTTT', 'ACCACGG', 'AATCCTA']"]},"metadata":{},"execution_count":68}]},{"cell_type":"code","source":[],"metadata":{"id":"8K2HmWnBk9lz"},"execution_count":null,"outputs":[]}]}