{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyMionrOOu8PAqNIbk7n/lvu"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["## WHERE IN THE GENOME DOES REPLICATION BEGIN?"],"metadata":{"id":"Xon0hCESw6f-"}},{"cell_type":"markdown","source":["#### PERCULIAR STATISTICS OF FORWARD AND REVERSE HALF-STRANDS 1.3"],"metadata":{"id":"wNS7xO8ExGTL"}},{"cell_type":"code","source":["# SkewProblem\n","# INPUT: Text string of genome snippet\n","# OUTPUT: List of integers representing the C-G value of the string at each index\n","\n","def Skew(genome:str) -> list[int]:\n","  count = 0\n","  skew = []\n","  # checking if we have any values in our string\n","  if len(genome) <= 1:\n","    return 0\n","  skew.append(count)\n","  for nucleotide in genome:\n","    # if the nucleotide is G, we add 1 to our skew value\n","    if nucleotide == 'G':\n","      count += 1 # the count value is now updated for the next appending\n","      skew.append(count)\n","    # if the nucleotide is C, we subtract 1 from our skew value\n","    if nucleotide == 'C':\n","      count -= 1\n","      skew.append(count)\n","    # we still need to append the count if C or G are not present.\n","    if nucleotide == \"A\" or nucleotide == \"T\":\n","      skew.append(count)\n","  return skew"],"metadata":{"id":"U09yU_VPxUJz","executionInfo":{"status":"ok","timestamp":1740086366909,"user_tz":300,"elapsed":31,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":5,"outputs":[]},{"cell_type":"code","source":["# testing a given string\n","genome = \"GAGCCACCGCGATA\"\n","answer = (Skew(genome))\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"FBtQ4ofTVape","executionInfo":{"status":"ok","timestamp":1740086366910,"user_tz":300,"elapsed":4,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"3835dde5-ede5-4032-8253-dba7f36cabf7"},"execution_count":6,"outputs":[{"output_type":"stream","name":"stdout","text":["[0, 1, 1, 2, 1, 0, 0, -1, -2, -1, -2, -1, -1, -1, -1]\n"]}]},{"cell_type":"code","source":["display(\" \".join(map(str, answer)))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":35},"id":"N1uB_aBKWgHM","executionInfo":{"status":"ok","timestamp":1740086366942,"user_tz":300,"elapsed":30,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"6401527b-4f24-4796-fe36-d6771c22374a"},"execution_count":7,"outputs":[{"output_type":"display_data","data":{"text/plain":["'0 1 1 2 1 0 0 -1 -2 -1 -2 -1 -1 -1 -1'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{}}]},{"cell_type":"markdown","source":["Time to find the minimum value within our Skew List to find the turning point in our larger Skew Graph"],"metadata":{"id":"d5lccBqmbKfW"}},{"cell_type":"code","source":["# MinSkew\n","# INPUT: Genome text string\n","# OUTPUT: List of ints that are the indices minimizing the Skew values\n","\n","def MinSkew(genome:str) -> list[int]:\n","  skewList = Skew(genome)\n","  minValue= min(skewList)\n","  minSkew = []\n","\n","  # now we will cycle through skewList and find the indices that have the minVal\n","  for i in range(len(skewList)):\n","    if skewList[i] == minValue:\n","      minSkew.append(i)\n","\n","\n","  return minSkew"],"metadata":{"id":"IPFGudK-bq1Q","executionInfo":{"status":"ok","timestamp":1740086366943,"user_tz":300,"elapsed":1,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":8,"outputs":[]},{"cell_type":"code","source":["# using given test sample\n","genome = \"TAAAGACTGCCGAGAGGCCAACACGAGTGCTAGAACGAGGGGCGTAAACGCGGGTCCGAT\"\n","print(MinSkew(genome))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"a9SaxDINlfpu","executionInfo":{"status":"ok","timestamp":1740086366947,"user_tz":300,"elapsed":3,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"95f30bfb-618d-4508-ba9a-cae695d4fefb"},"execution_count":9,"outputs":[{"output_type":"stream","name":"stdout","text":["[11, 24]\n"]}]},{"cell_type":"markdown","source":["#### SOME HIDDEN MESSAGES ARE MORE ELUSIVE THAN OTHERS 1.4"],"metadata":{"id":"CYNJDUy5pPsa"}},{"cell_type":"markdown","source":["We will be looking at the DNA sequences that are SIMILAR to our target pattern since there are Single Nucleotide Point mutations that can occur within the genome. This will allow our algorithm to be more robust and find sequences that are close our target sequence.<br>\n","\n","We can compute this different by calculating the Hamming Distance between our target sequence and the sequence found in the genome. This calculation is done by adding the number of times p[i] does NOT equal q[i]."],"metadata":{"id":"LeKCSl9FpYo7"}},{"cell_type":"code","source":["# HAMMING DISTANCE\n","# INPUT: Two strings of equal length\n","# OUTPUT: The hamming distance between the two strings\n","\n","def HammingDistance(p: str, q:str) -> int:\n","  hd = 0\n","  for i in range(len(p)):\n","    if p[i] != q[i]:\n","      hd += 1\n","  return hd"],"metadata":{"id":"ryCIkxx8p9CG","executionInfo":{"status":"ok","timestamp":1740086366948,"user_tz":300,"elapsed":1,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":10,"outputs":[]},{"cell_type":"markdown","source":["Now we can have the necessary functions to solve the \"Approximate Pattern Matching\" problem. We will search for patterns that have a Hamming Distance of at most \"d\" from our desired pattern."],"metadata":{"id":"IoZ24r9CzVef"}},{"cell_type":"code","source":["# APPROX. Pattern Matching\n","# INPUT: Strings \"Pattern\" and \"Text\" along with integer \"d\"\n","# OUTPUT: All starting positions where Pattern appears with at most \"d\" mismatches as a subtring in Text\n","\n","def ApproxPatternMatching(genome:str, pattern:str, d:int) -> list[int]:\n","  patterns = []\n","  n = len(genome)\n","  k = len(pattern)\n","\n","  for i in range(n-k+1):\n","    subString = genome[i:k+i]\n","    if HammingDistance(subString,pattern) <= d:\n","      patterns.append(i)\n","\n","  return patterns"],"metadata":{"id":"zsHojD-_zblg","executionInfo":{"status":"ok","timestamp":1740086366949,"user_tz":300,"elapsed":0,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":11,"outputs":[]},{"cell_type":"code","source":["# test with given test set\n","pattern = \"ATTCTGGA\"\n","genome = \"CGCCCGAATCCAGAACGCATTCCCATATTTCGGGACCACTGGCCTCCACGGTACGGACGTCAATCAAAT\"\n","d = 3\n","\n","answer = ApproxPatternMatching(genome,pattern,d)\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"B-syqVVr0qAv","executionInfo":{"status":"ok","timestamp":1740086366979,"user_tz":300,"elapsed":29,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"1bd43420-f6e5-417b-cbbb-228685a4a006"},"execution_count":12,"outputs":[{"output_type":"stream","name":"stdout","text":["[6, 7, 26, 27]\n"]}]},{"cell_type":"code","source":["# # file provided by cogniterra for testing\n","# file_path = \"/content/sample_data/dataset_30278_4.txt\"\n","\n","# # reading each line and deleting the space between each in the txt file:\n","# with open(file_path,\"r\") as file:\n","#   lines = file.readlines()\n","#   # lines here is a whole string, with spaces separating each\n","#   lines = [line.rstrip(\"\\n\") for line in lines]\n","\n","#   print(lines)"],"metadata":{"id":"F7o2zDGl04tW","executionInfo":{"status":"ok","timestamp":1740086366980,"user_tz":300,"elapsed":9,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":13,"outputs":[]},{"cell_type":"code","source":["# pattern = lines[0]\n","# genome = lines[1]\n","# d = int(lines[2])\n","\n","# answer = ApproxPatternMatching(genome,pattern,d)\n","# new_answer = \" \".join(map(str,answer))"],"metadata":{"id":"xLTjn9Vi1ilt","executionInfo":{"status":"ok","timestamp":1740086366981,"user_tz":300,"elapsed":9,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":14,"outputs":[]},{"cell_type":"markdown","source":["We now want to adjust our \"FrequentWords\" problem of looking for most frequent k-mers within a genome, to also searching for k-mers with at most d-mismatches. This will make our searching algorithm more robust and account for this Single Nucleotide Polymorphisms, or rather single character changes within our given Text string."],"metadata":{"id":"UROJVxdk_x2B"}},{"cell_type":"code","source":["# ApproxPatternCount\n","# INPUT: Genome and Pattern as strings and \"d\" as integer\n","# OUTPUT: A list of the most frequent k-mers that have at most d-mismatches from k-mers in the given Genome\n","\n","def ApproxPatternCount(genome: str, pattern: str, d: int) -> list[str]:\n","  count = 0\n","  n = len(genome)\n","  k = len(pattern)\n","\n","  # loop through the entire given genome.\n","  for i in range(n-k+1):\n","    # subString is the pattern we are testing against our given pattern.\n","    subString = genome[i:i+k]\n","    # if the number of mismatches (ie. the Hamming Distance)\n","    # is less than d, we can add that pattern to our freqPatterns list.\n","    if HammingDistance(subString,pattern) <= d:\n","      count += 1\n","\n","  return count"],"metadata":{"id":"CMRB2cCCAPoC","executionInfo":{"status":"ok","timestamp":1740086366982,"user_tz":300,"elapsed":9,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":15,"outputs":[]},{"cell_type":"code","source":["genome = \"AACAAGCTGATAAACATTTAAAGAG\"\n","pattern = \"AAAAA\"\n","\n","answer = ApproxPatternCount(genome, pattern, 2)\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"490J7zzHCQIw","executionInfo":{"status":"ok","timestamp":1740086366983,"user_tz":300,"elapsed":10,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"748264b1-f9c9-46d5-c7d2-5bf4dc525c8e"},"execution_count":16,"outputs":[{"output_type":"stream","name":"stdout","text":["11\n"]}]},{"cell_type":"code","source":["# dataset = \"/content/sample_data/dataset_30278_7.txt\"\n","# with open(dataset, \"r\") as file:\n","#   lines = file.readlines()\n","#   # using this technique because lines is a list, not just a string\n","#   new_lines = [string.replace(\"\\n\", \"\") for string in lines]\n","#   print(new_lines)"],"metadata":{"id":"cQt3Ov5pCcBp","executionInfo":{"status":"ok","timestamp":1740086366994,"user_tz":300,"elapsed":11,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":17,"outputs":[]},{"cell_type":"code","source":["# pattern = new_lines[0]\n","# genome = new_lines[1]\n","# d = int(new_lines[2])\n","\n","# answer = ApproxPatternCount(genome,pattern, d)\n","# print(answer) # -> CORRECT!"],"metadata":{"id":"bX9msWXJc2C4","executionInfo":{"status":"ok","timestamp":1740086366995,"user_tz":300,"elapsed":11,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":18,"outputs":[]},{"cell_type":"markdown","source":["Now we can apply our logic for out Approx. Count function to an updated \"Frequent Words\" algorithm. In this new approach, we will be also looking for the patterns that are at most d mismatches away from the patterns found in the given genome string."],"metadata":{"id":"XV_nfA0UfUQv"}},{"cell_type":"code","source":["\"\"\"\n","Pseudo Code for \"ApproxFreqWords\":\n","FrequentWordsWithMismathces(Text, k, d)\n","    Patterns <- Array of string with length 0\n","    freqMap <- Empty Map\n","    n <- Length of given Text string\n","    FOR i <- 0 to n-k\n","        pattern <- Text(i:k)\n","        neighborhood <- Neighbors(Pattern, d)\n","        FROM j <- 0 to length of neighborhood  - 1\n","            neighbot <- Neighborhood[j]\n","            IF freqMap[neighbor] doesn't exist\n","                freqMap[neighbor] <- 1\n","            ELSE\n","                freqMap[nrighbor] <- freqMap[neighbor] + 1\n","\n","    m <- MaxMap(freqMap)\n","    FOR every pattern in freqMa\n","        IF freqMap[pattern] <- m\n","            append pattern to Patterns\n","\n","    return Patterns\n","\"\"\""],"metadata":{"id":"fy_KehuinVL1","colab":{"base_uri":"https://localhost:8080/","height":87},"executionInfo":{"status":"ok","timestamp":1740086366996,"user_tz":300,"elapsed":4,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"3bb8264a-9874-4f02-9fef-3cf6d5927e39"},"execution_count":19,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'\\nPseudo Code for \"ApproxFreqWords\":\\nFrequentWordsWithMismathces(Text, k, d)\\n    Patterns <- Array of string with length 0\\n    freqMap <- Empty Map\\n    n <- Length of given Text string\\n    FOR i <- 0 to n-k\\n        pattern <- Text(i:k)\\n        neighborhood <- Neighbors(Pattern, d)\\n        FROM j <- 0 to length of neighborhood  - 1\\n            neighbot <- Neighborhood[j]\\n            IF freqMap[neighbor] doesn\\'t exist\\n                freqMap[neighbor] <- 1\\n            ELSE\\n                freqMap[nrighbor] <- freqMap[neighbor] + 1\\n\\n    m <- MaxMap(freqMap)\\n    FOR every pattern in freqMa\\n        IF freqMap[pattern] <- m\\n            append pattern to Patterns\\n\\n    return Patterns\\n'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":19}]},{"cell_type":"code","source":["# first we need to make our Neighborhood function.\n","# NEIGHBORS FUNCTION\n","# Recursive Function that takes an input pattern (string) and outputs a collection of neighbor sequences that have a hamming distance of at most d.\n","# INPUT: A string Pattern and an integer d (for the max hamming distance)\n","# OUTPUT: A collection of strings Neighbors(pattern,d)\n","\n","def Neighbors(pattern: str, d: int) -> list[str]:\n","  nucleotides = ['G', 'T', 'A','C']\n","  # for HammingDistance to be 0, neighbor == pattern\n","  if d == 0:\n","    return [pattern]\n","  # If length pattern = 1, then it is a single nucleotide, so the neightbors will just be the other nucleotides\n","  if len(pattern) == 1:\n","    return nucleotides\n","  neighborhood = [] # creating empty set\n","  SuffixPattern = pattern[1:] # Dropping the first symbol in pattern to get the SuffixPattern\n","  SuffixNeighbors = Neighbors(SuffixPattern, d) # Recursion, will keep calling function until base case (len(pattern) == 1)\n","\n","  for text in SuffixNeighbors:\n","    if HammingDistance(SuffixPattern, text) < d: # <d = there can be more mismatching allowed\n","      for nucleotide in nucleotides:\n","        # we now add a nucleotide as each possible neigbor to pattern\n","        neighbor = nucleotide + text\n","        neighborhood.append(neighbor)\n","    else: # HammingDistance is equal to d, thus no more mismatching allowed and the prefix needs to be the prefix of the original pattern\n","      neighbor = pattern[0] + text # SuffixPattern[0] is the first symbol in SuffixPattern\n","      neighborhood.append(neighbor)\n","\n","  return neighborhood"],"metadata":{"id":"IvLSSOanNchE","executionInfo":{"status":"ok","timestamp":1740086367028,"user_tz":300,"elapsed":34,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":20,"outputs":[]},{"cell_type":"code","source":["# TESTING NEIGHBORS()\n","\n","pattern = \"ACG\"\n","d = 1\n","\n","neighborhood = Neighbors(pattern, d)\n","\n","print(neighborhood)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"hDl8cktoOWmw","executionInfo":{"status":"ok","timestamp":1740086367029,"user_tz":300,"elapsed":4,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"898cb653-706e-4836-a451-a31a8a4de9cd"},"execution_count":21,"outputs":[{"output_type":"stream","name":"stdout","text":["['AGG', 'ATG', 'AAG', 'GCG', 'TCG', 'ACG', 'CCG', 'ACT', 'ACA', 'ACC']\n"]}]},{"cell_type":"code","source":["answer = [\"ACG\", \"ACT\", \"ACA\", \"ACC\", \"ATG\" ,\"AGG\", \"AAG\" ,\"TCG\", \"GCG\", \"CCG\"]\n","check = all(pattern in neighborhood for pattern in answer)\n","\n","if check == True:\n","  print(\"Elements in neighborhood are all in answer\")\n","else:\n","  print(\"Not all elements in neighborhood are in answer\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fC1JhVdpOvEN","executionInfo":{"status":"ok","timestamp":1740086388788,"user_tz":300,"elapsed":41,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"810f3aea-9583-400e-c30b-831a9fd24556"},"execution_count":23,"outputs":[{"output_type":"stream","name":"stdout","text":["Elements in neighborhood are all in answer\n"]}]},{"cell_type":"code","source":["# recreating our MaxMap function from part one of this chapter:\n","# INPUT: a frequecy table as a map\n","# OUTPUT: The maximum value within a frequency table\n","\n","def MaxMap(freqMap: map) -> int:\n","  max = 0\n","  for pattern in freqMap:\n","    if freqMap[pattern] > max:\n","      max = freqMap[pattern]\n","  return max"],"metadata":{"id":"ZBVViiPbPJgq","executionInfo":{"status":"ok","timestamp":1740086497718,"user_tz":300,"elapsed":1,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":24,"outputs":[]},{"cell_type":"code","source":["# Now we will find the most frequent k-mer WITH mismatches\n","# we want to return the kmer pattern, not the ones with mismatches\n","\n","# INPUT: A string text (genome) and integers k for the length of the pattern and d for max Hamming Distance\n","# OUTPUT: A list of most frequence k-mers with up to d-mismatches\n","def FrequentWordsWithMismatches(text: str, k: int, d: int) -> list[str]:\n","  patterns = []\n","  freqMap = {}\n","  n = len(text)\n","\n","  for i in range(n-k+1):\n","    pattern = text[i:i+k]\n","    neighborhood = Neighbors(pattern, d)\n","    for j in range(len(neighborhood)):\n","      neighbor = neighborhood[j]\n","      if neighbor not in freqMap:\n","        freqMap[neighbor] = 1\n","      else:\n","        freqMap[neighbor] += 1\n","\n","  # m = MaxMap(freqMap) # -> can run a bit faster if we use Python's buily in \"max\" function\n","  m = max(freqMap.values())\n","  for pattern in freqMap:\n","    if freqMap[pattern] == m:\n","      patterns.append(pattern)\n","\n","\n","  return patterns"],"metadata":{"id":"EyHnxiuYtLf2","executionInfo":{"status":"ok","timestamp":1740086676300,"user_tz":300,"elapsed":8,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":27,"outputs":[]},{"cell_type":"code","source":["test_text = \"AGGT\"\n","k = 2\n","d = 1\n","\n","answer = FrequentWordsWithMismatches(test_text, k , d)\n","\n","print(answer)\n","\n","# first attempt was incorrect.\n","# after writing the algorithm on paper, found that I did no account for the zero based indexing\n","# fixed the issue by adding +1 to the range for i in the first \"for\" loop."],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"hnCyZNnqPoYq","executionInfo":{"status":"ok","timestamp":1740086744253,"user_tz":300,"elapsed":8,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"c247198e-f568-43f2-ec12-fbf960e30241"},"execution_count":29,"outputs":[{"output_type":"stream","name":"stdout","text":["['GG']\n"]}]},{"cell_type":"code","source":["# # testing our function with the given text file:\n","# dataset = \"/content/sample_data/dataset_30278_10.txt\"\n","\n","# with open(dataset, 'r') as file:\n","#   lines = file.readlines()\n","#   new_lines = [string.replace(\"\\n\",\"\") for string in lines]\n","#   new_lines.extend(new_lines[1].split(\" \"))\n","#   new_lines.pop(1)\n","\n","#   print(new_lines)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"GeP2cNrKPtr4","executionInfo":{"status":"ok","timestamp":1740088231159,"user_tz":300,"elapsed":3,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"4074cb68-6cb1-47f5-aea3-19de2afde494"},"execution_count":49,"outputs":[{"output_type":"stream","name":"stdout","text":["['TGCATAGGGTATAGTGCATGCATGCATAGGGTACACTAGTAGCGCTGCATGCAGGTATAGTGCAGGTATGCATGCATGCATAGCACGGTACGCCGCCGCTGCACGCTGCACACTAGTGCACACTAGCGCGGTAGGTATGCATAGTGCAGGTACACCGCGGTACACCACCACCGCGGTACACTAGTGCATAGGGTAGGTATAGTGCAGGTACGCCACCGCTAGCACCACGGTATGCACGCCGCCGCGGTACACGGTATAGGGTATAGTGCATAGTAGCACGGTAGGTACGCCGCTGCACGCTAGCACTAGCGCTAGGGTACACCGCTAGCACCACTAGGGTAGGTATGCATGCATAGCGCTGCATAGCGCTGCA', '7', '3']\n"]}]},{"cell_type":"code","source":["# genome = new_lines[0]\n","# k = int(new_lines[1])\n","# d = int(new_lines[2])\n","\n","# answer = FrequentWordsWithMismatches(genome,k,d)\n","# print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"8p2dtFeIQykQ","executionInfo":{"status":"ok","timestamp":1740088299093,"user_tz":300,"elapsed":555,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"4cdd93c4-7353-402b-9a89-62c0aa7d1018"},"execution_count":51,"outputs":[{"output_type":"stream","name":"stdout","text":["['GGCACGG']\n"]}]},{"cell_type":"markdown","source":["Now that we have a working function for finding the most frequent k-mers with at most d mismatches. We can tackle finding the most Frequent k-mers that take into account mismatches **AND** their reverse complement."],"metadata":{"id":"1Phgx3TEWkiH"}},{"cell_type":"code","source":["# Function for finding the Reverse Complement of a string\n","\n","# Insert your reverse_complement function here, along with any subroutines you need\n","# ReverseComplement takes a string pattern as input, creates a complement strand then reverses it\n","# INPUT: String Pattern\n","# OUTPUT: Reverse Complement of given Pattern\n","def ReverseComplement(pattern: str) -> str:\n","    complement = Complement(pattern)\n","    reverse_complement = complement[::-1]\n","\n","    return reverse_complement\n","\n","\n","\n","# Complement takes a pattern string and returns the complementary strand of that pattern string\n","# INPUT: String pattern\n","# OUTPUT: Complement of string pattern\n","def Complement(pattern: str) -> str:\n","    # creating an empty string\n","    complement = \"\"\n","    for i in range(len(pattern)):\n","        # checking the nucleotide and giving the concatenating the complement\n","        # Checking A to T, C to G and vice versa\n","        if pattern[i] == \"A\":\n","            complement += \"T\"\n","        elif pattern[i] == \"T\":\n","            complement += \"A\"\n","        elif pattern[i] == \"C\":\n","            complement += \"G\"\n","        elif pattern[i] == \"G\":\n","            complement += \"C\"\n","\n","    return complement"],"metadata":{"id":"nrx1qqjxXYJW","executionInfo":{"status":"ok","timestamp":1740088659956,"user_tz":300,"elapsed":41,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":52,"outputs":[]},{"cell_type":"code","source":["# VERSION 1.1\n","# Frequent words problem WITH finding the reverse complement\n","# INPUT: A DNA string Text , integer k for length of pattern and integer d for max Hamming Distance.\n","# OUPUT: All k-mers Pattern maximizing the sum Count(Text, Pattern) + Count(Text, Pattern_RC) over all possible k-mers\n","\n","def FrequentWordsWithMismatchesReverseComplements(text: str, k: int, d: int) -> list[str]:\n","  patterns = []\n","  freqMap = {}\n","  n = len(text)\n","\n","  # we will make a rc_neighborhood to search for possible neighbors of reverse complements of the patterns we find\n","  # we will then append the regular neighborhood and rc_neighborhood to find the most frequent k-mers in this bigger neighborhood\n","  for i in range(n-k+1):\n","    pattern = text[i:i+k]\n","    neighborhood = Neighbors(pattern, d)\n","    rc_pattern = ReverseComplement(pattern)\n","    rc_neighborhood = Neighbors(rc_pattern, d)\n","    neighborhood = neighborhood + rc_neighborhood\n","    for j in range(len(neighborhood)):\n","      neighbor = neighborhood[j]\n","      if neighbor not in freqMap:\n","        freqMap[neighbor] = 1\n","      else:\n","        freqMap[neighbor] += 1\n","\n","  m = MaxMap(freqMap)\n","  for pattern in freqMap:\n","    if freqMap[pattern] == m:\n","      patterns.append(pattern)\n","\n","\n","  return patterns\n"],"metadata":{"id":"8vITTYJIXErw","executionInfo":{"status":"ok","timestamp":1740089306959,"user_tz":300,"elapsed":3,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":77,"outputs":[]},{"cell_type":"code","source":["test_sequence = \"ACGTTGCATGTCGCATGATGCATGAGAGCT\"\n","k = 4\n","d = 1\n","\n","answer = FrequentWordsWithMismatchesReverseComplements(test_sequence, k, d)\n","print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YHYPmXEoXk93","executionInfo":{"status":"ok","timestamp":1740089323165,"user_tz":300,"elapsed":6,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"dc83fd75-3f6f-45a5-d197-6ae6e54b0594"},"execution_count":79,"outputs":[{"output_type":"stream","name":"stdout","text":["['ATGT', 'ACAT']\n"]}]},{"cell_type":"code","source":["# # testing our function with the given text file:\n","# dataset = \"/content/sample_data/dataset_30278_12.txt\"\n","\n","# with open(dataset, 'r') as file:\n","#   lines = file.readlines()\n","#   new_lines = [string.replace(\"\\n\",\"\") for string in lines]\n","#   new_lines.extend(new_lines[1].split(\" \"))\n","#   new_lines.pop(1)\n","\n","#   print(new_lines)\n","\n","new_lines = ['GATCGGAGATCGTCTACATCGGATCTGAACATCTGAACGATCGACATCTACAACAACGATCTGAACAACTCTGAACACATCTACTCTACAGAGATCTACTCGTCGTCGTCTACATCTTCTACTCGGAGAGAACTCGACACTCTGATCTTCTTCGTCGGATCTACATCGACATCTACAGATCTACATCTACAGAACTCGGATCGACACAACAACGAGATCGTCTTCGGAACATCTTCTACAAC', '7', '3']"],"metadata":{"id":"wqco3ivfZ7At","executionInfo":{"status":"ok","timestamp":1740091979677,"user_tz":300,"elapsed":3,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":89,"outputs":[]},{"cell_type":"code","source":["genome = new_lines[0]\n","k = int(new_lines[1])\n","d = int(new_lines[2])\n","\n","answer = FrequentWordsWithMismatchesReverseComplements(genome,k,d)"],"metadata":{"id":"q7A8lW8vjDwT","executionInfo":{"status":"ok","timestamp":1740091982855,"user_tz":300,"elapsed":685,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":90,"outputs":[]},{"cell_type":"code","source":["print(answer)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"QLWZqp6GjqY5","executionInfo":{"status":"ok","timestamp":1740091982865,"user_tz":300,"elapsed":5,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"2877acb1-59b2-4ced-dddc-d9ede681164c"},"execution_count":91,"outputs":[{"output_type":"stream","name":"stdout","text":["['TAGATAT', 'ATATCTA']\n"]}]},{"cell_type":"markdown","source":["Although we are looking for 9-mers that can be related to the location of the Origin of Replication. We find other 9-mers clusterd in various bacterial genomes because there are many hidden messages within the genome that we do not as of yet know what they are specifically for. <br/>\n","\n","The main take away from this is that the existing methods for finding the origin of replication is still imperfect and at times inconclusive. However, this work is not in vain, as it narrows the window of research for biologists and gives a great starting point for further experimentation and observation."],"metadata":{"id":"rAnaq2likgAF"}},{"cell_type":"markdown","source":["#### EPILOGUE: COMPLICATIONS IN ORI PREDICITONS 1.6"],"metadata":{"id":"xXdb0kOZl2XX"}},{"cell_type":"markdown","source":["- Skew diagram is only a rough estrimation of where the Ori can be.\n","- The Skew Diagram is not always a clear diagram.\n"],"metadata":{"id":"1xyJVurqnd5E"}},{"cell_type":"markdown","source":["**Now we will attempt to find the DnaA Box of *Salmonella Enterica*.**"],"metadata":{"id":"Wil4YskjqVer"}},{"cell_type":"code","source":["se_dataset = \"/content/sample_data/Salmonella_enterica.txt\"\n","with open(se_dataset,'r') as file:\n","  next(file)\n","  s_enterica = file.read()\n","  s_enterica = s_enterica.replace(\"\\n\",\"\")"],"metadata":{"id":"K6l7RT4bqhkE","executionInfo":{"status":"ok","timestamp":1740093879922,"user_tz":300,"elapsed":20,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}}},"execution_count":97,"outputs":[]},{"cell_type":"code","source":["s_enterica"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":122,"output_embedded_package_id":"1vCBnwTFfrBQPzRGA_1KLR-qUhpwbXenL"},"id":"y0tfWUZwrJuU","executionInfo":{"status":"ok","timestamp":1740093881449,"user_tz":300,"elapsed":837,"user":{"displayName":"Adonis Miliano","userId":"11068781965103758971"}},"outputId":"5205f851-0f59-4ac5-a9e0-d0f6d650a51a"},"execution_count":98,"outputs":[{"output_type":"display_data","data":{"text/plain":"Output hidden; open in https://colab.research.google.com to view."},"metadata":{}}]},{"cell_type":"code","source":["# we will first look for the minimum skew\n","# with the location of the minium skew, we will start our FrequentWords problem\n","# FreqWords will have a window (L) of 1000, K of 9 and a d of 2.\n","# we will run FreqWords 3 times, once at the min_skew, then once 500bp before, and then 500bp after."],"metadata":{"id":"nW8c7BmHrKaa"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"gMuTngGmuVGs"},"execution_count":null,"outputs":[]}]}